---
title: "Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# Overview

In this tutorial, we will illustrate how to recover spatial ecotypes (SEs) from single-cell gene expression data, including scRNA-seq and single-cell-scale spatial transcriptomics datasets generated by platforms such as MERSCOPE and Xenium.

__First load required packages for this vignette__

```{r lib}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(parallel))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(googledrive))
library(SpatialEcoTyper)
```

# Data preparation

The recovery process requires a normalized gene expression matrix and a vector of cell type annotations. 

<details open><summary><strong>Starting from a Seurat object</strong></summary>

A seurat object for the demo data can be accessed from  [WU2161_seurat_obj.rds](https://drive.google.com/open?id=11ORWQxxWCNUtceEDwu2IyBpDMa6P-bxS&usp=drive_fs).

```{r downloadscrna, eval=FALSE}
drive_deauth() # Disable Google sign-in requirement
drive_download(as_id("11ORWQxxWCNUtceEDwu2IyBpDMa6P-bxS"), "WU2161_seurat_obj.rds", overwrite = TRUE)
```

First, load the seurat object into R:
```{r loadscrna}
# Load the seurat object
obj <- readRDS("WU2161_seurat_obj.rds")
obj

# Cell type annotations
unique(obj$CellType)
```

Next, normalize the gene expression data using [SCTransform](https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2) or [NormalizeData](https://satijalab.org/seurat/reference/normalizedata). Here, we are normalizing the data using `SCTransform`. We recommend to install the `glmGamPoi` package for faster computation.

```{r SCTransform, eval = TRUE}
if(!"glmGamPoi" %in% installed.packages()){
  BiocManager::install("glmGamPoi")
}
obj <- SCTransform(obj, verbose = FALSE)
```

Then, extract the normalized gene expression matrix and cell type annotations from the Seurat object.

```{r extract}
seurat_version = as.integer(gsub("\\..*", "", as.character(packageVersion("SeuratObject"))))
if(seurat_version<5){
  normdata <- GetAssayData(obj, "data")
}else{
  normdata <- obj[["SCT"]]$data
}
head(normdata[, 1:5])

ctann = obj$CellType
head(ctann)
table(ctann)
```

</details>

<details><summary><strong>Starting from Sparse matrix</strong></summary>
Sparse matrix in the .mtx format can be imported using the `ReadMtx` function from the `Seurat` package. The demo data can be accessed from  [WU2161](https://drive.google.com/open?id=178yvBUNNAAJ8TGtXmvjplVz_Ev7V0WxA&usp=drive_fs). The cell type annotations are available in [WU2161_celltype_ann.tsv](https://drive.google.com/open?id=17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ&usp=drive_fs).

```{r loadsparse, eval=FALSE}
## Load the gene expression matrix
scdata = ReadMtx(mtx = "matrix.mtx", features = "features.tsv", 
                 cells = "barcodes.tsv", feature.column = 1, cell.column = 1)

## Normalize the data
normdata = NormalizeData(scdata)
head(normdata[, 1:5])
## Load cell type annotation
ctann = read.table("WU2161_celltype_ann.tsv", sep = "\t", 
                   header = TRUE, row.names = 1)
ctann = ctann[match(colnames(normdata), rownames(ctann)), 1]
table(ctann)
```

</details>


<details><summary><strong>Starting from tab-delimited files</strong></summary>

Tab-delimited files can be loaded into R using the `fread` function from the `data.table` package. The TSV file for the expression matrix can be accessed from  [WU2161_counts.tsv](https://drive.google.com/open?id=17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei&usp=drive_fs) and the cell type annotations are available in [WU2161_celltype_ann.tsv](https://drive.google.com/open?id=17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ&usp=drive_fs).

```{r downloadtsv, eval=FALSE}
## Download data from google drive
drive_download(as_id("17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei"), "WU2161_counts.tsv", overwrite = TRUE)
drive_download(as_id("17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ"), "WU2161_celltype_ann.tsv", overwrite = TRUE)
```


```{r loadtsv, eval=TRUE}
## Load the gene expression matrix
scdata = fread("WU2161_counts.tsv", sep = "\t", data.table = FALSE)
rownames(scdata) = scdata[, 1] ## Set the first column as row names
scdata = scdata[, -1] ## Drop the first column

## Normalize the data
normdata = NormalizeData(scdata)
head(normdata[, 1:5])

## Load cell type annotation
ctann = read.table("WU2161_celltype_ann.tsv", sep = "\t", 
                   header = TRUE, row.names = 1)
ctann = ctann[match(colnames(normdata), rownames(ctann)), 1]
table(ctann)
```

</details>


# SE recovery

The [RecoverSE](../reference/RecoverSE.html) function will be used to assign single cells into SEs. Users can either use default model to recover predefined SEs or use custom model to recover newly defined SEs.

**Note**: Cell type annotation (`celltypes` argument) is required for SE recovery using [RecoverSE](../reference/RecoverSE.html).

<details open><summary><strong>Using default models</strong></summary>

The default NMF models were trained on discovery MERSCOPE data, encompassing five cancer types: melanoma, and four carcinomas. These models are tailored to nine distinct cell types: B cells, CD4+ T cells, CD8+ T cells, NK cells, plasma cells, macrophages, dendritic cells, fibroblasts, and endothelial cells. Each model facilitates the recovery of SEs from single-cell datasets, allowing for cell-type-specific SE analysis.

For SE recovery, the cells in the query data should be grouped into one of "B", "CD4T", "CD8T", "NK", "Plasma", "Macrophage", "DC", "Fibroblast", and "Endothelial", case sensitive. All the other cell types will be considered non-SE compartments. 

```{r scrna_recov}
sepreds <- RecoverSE(normdata, celltypes = ctann)
head(sepreds)
```

</details>

<details><summary><strong>Using custom models</strong></summary>

To use custom models, users should first develop recovery models following the tutorial [Development of SE Recovery Models](TrainRecoveryModels.html). The resulting models can be used for SE recovery. An example model is available at [SE_Recovery_W_list.rds](https://drive.google.com/open?id=171WaAe49babYB85Cn1FcoNNE-lzYp1T_&usp=drive_fs).

Download the example models
```{r downloadWs, eval=FALSE}
drive_download(as_id("171WaAe49babYB85Cn1FcoNNE-lzYp1T_"), "SE_Recovery_W_list.rds", overwrite = TRUE)
```

Load the custom models
```{r viewWs}
Ws <- readRDS("SE_Recovery_W_list.rds")
names(Ws) ## named list of W matrices
head(Ws[[1]]) ## feature by SE matrix
```

Using custom models for SE recovery by specifying the `Ws` argument.
```{r scrna_recov2, eval=TRUE}
sepreds <- RecoverSE(normdata, celltypes = ctann, Ws = Ws)
head(sepreds)
```
</details>

# Session info

The session info allows users to replicate the exact environment and identify potential discrepancies in package versions or configurations that might be causing problems.

```{r session}
sessionInfo()
```
