<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial ecotype analaysis • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Spatial ecotype analaysis">
<meta name="description" content="SpatialEcoTyper is an unsupervised machine learning framework for discovering spatial ecotypes (multi-cellular communities) from single-cell spatial data.">
<meta property="og:description" content="SpatialEcoTyper is an unsupervised machine learning framework for discovering spatial ecotypes (multi-cellular communities) from single-cell spatial data.">
<meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p align="center">
<img width="800" src="logo.png"></p>
<div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="deciphering-spatial-ecotypes-from-spatial-single-cell-and-bulk-transcriptomic-data">Deciphering spatial ecotypes from spatial, single-cell, and bulk transcriptomic data<a class="anchor" aria-label="anchor" href="#deciphering-spatial-ecotypes-from-spatial-single-cell-and-bulk-transcriptomic-data"></a>
</h1>
</div>
<p><strong>Spatial EcoTyper</strong> is a versatile framework designed for the systematic identification of spatial cellular communities, termed spatial ecotypes, from single-cell spatial transcriptomics data. In addition, it provides unified methods for the recovery of spatial ecotypes across multiple data modalities, including spatial transcriptomics, single-cell RNA-seq, and bulk transcriptomic datasets.</p>
<div class="section level2">
<h2 id="overview-of-spatial-ecotyper">Overview of Spatial EcoTyper<a class="anchor" aria-label="anchor" href="#overview-of-spatial-ecotyper"></a>
</h2>
<p><strong>Spatial EcoTyper</strong> is available as an R package, with comprehensive documentation accessible at <a href="https://digitalcytometry.github.io/spatialecotyper" class="uri">https://digitalcytometry.github.io/spatialecotyper</a>.</p>
<p>We provide five comprehensive tutorials illustrating the functionalities included in the SpatialEcoTyper R package. The first tutorial demonstrates how to identify spatial ecotypes from a single-cell spatial transcriptomics data. The second demonstrates how to identified conserved spatial ecotypes across multiple samples. The third demonstrates how to develop NMF models for recovery of spatial ecotypes from unseen data. The remaining tutorials introduce how to recover spatial ecotypes from single-cell (scRNA-seq or single-cell spatial transcriptomics) and bulk (bulk RNA-seq or Visium) gene expression profiles.</p>
<ul>
<li>
<strong>Tutorial 1:</strong> <a href="https://digitalcytometry.github.io/spatialecotyper/articles/SingleSample.html">Discovery of Spatial Ecotypes from A Single Sample</a>
</li>
<li>
<strong>Tutorial 2:</strong> <a href="https://digitalcytometry.github.io/spatialecotyper/articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a>
</li>
<li>
<strong>Tutorial 3:</strong> <a href="https://digitalcytometry.github.io/spatialecotyper/articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a>
</li>
<li>
<strong>Tutorial 4:</strong> <a href="https://digitalcytometry.github.io/spatialecotyper/articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a>
</li>
<li>
<strong>Tutorial 5:</strong> <a href="https://digitalcytometry.github.io/spatialecotyper/articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a>
</li>
</ul>
<p>Data used in the tutorials can be obtained from <a href="https://drive.google.com/open?id=1En0pgY6_3_u8XK2hTSTbrA9Iouhvc4TD&amp;usp=drive_fs" class="external-link uri">https://drive.google.com/open?id=1En0pgY6_3_u8XK2hTSTbrA9Iouhvc4TD&amp;usp=drive_fs</a>.</p>
<p><strong>Note</strong>: <strong>Spatial EcoTyper</strong> depends extensively on Seurat for key processes like dimensionality reduction, UMAP embedding, clustering, and visualization. Initially developed using Seurat v4.3, the tool has been thoroughly tested and validated with Seurat v5. Although UMAP embeddings and clustering results show slight differences between Seurat v4 and v5, the overall consistency remains strong, ensuring that core biological insights are preserved across both versions.</p>
</div>
<div class="section level2">
<h2 id="system-requirements">System requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<p>This package is compatible with all operating systems and has been tested on the following platforms:</p>
<p>macOS: Big Sur, Monterey, Ventura, Sonoma, Sequoia (15.2) Linux: CentOS 7.2 and High-Performance Computing (HPC) clusters</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Spatial EcoTyper</strong> is available as an R package and can be installed via the <code>BiocManager</code> package directly from the R console.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Install dependencies</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"remotes"</span>, <span class="st">"Seurat"</span>, <span class="st">"NMF"</span>, <span class="st">"dplyr"</span>, <span class="st">"pals"</span>,</span>
<span>                       <span class="st">"data.table"</span>, <span class="st">"ComplexHeatmap"</span>, <span class="st">"googledrive"</span>, </span>
<span>                       <span class="st">"glmGamPoi"</span>, <span class="st">"immunogenomics/presto"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install SpatialEcoTyper</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"digitalcytometry/spatialecotyper"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
Troubleshooting dependency installation
</summary><ul>
<li>
<p>ERROR: dependency ‘GetoptLong’ is not available for package ‘ComplexHeatmap’</p>
<p>If the installation within R console fails, you can try installing the necessary packages via <code>conda install</code> or <code>mamba install</code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">conda</span> install bioconda::bioconductor-complexheatmap</span></code></pre></div>
</li>
<li>
<p>Failed to install ‘presto’ from GitHub: HTTP error 401. Bad credentials</p>
<p>To resolve this issue, you’ll need to authenticate using a personal access token (PAT). You can generate a GitHub personal access token following the <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens" class="external-link">GitHub’s documentation</a>. After generating the token, set it as an environment variable in R using the following code. Replace “YOUR_TOKEN” with your actual token:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set the token in your R environment:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>GITHUB_PAT<span class="op">=</span><span class="st">"YOUR_TOKEN"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Install the package from GitHub:</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"immunogenomics/presto"</span><span class="op">)</span></span></code></pre></div>
</li>
</ul></details><details><summary>
Install SpatialEcoTyper from source code
</summary><p>The source code of <strong>Spatial EcoTyper</strong> is available at <a href="https://github.com/digitalcytometry/spatialecotyper" class="external-link uri">https://github.com/digitalcytometry/spatialecotyper</a>. After downloading the package, you can install it from the source code using the command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"SpatialEcoTyper.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="contribution">Contribution<a class="anchor" aria-label="anchor" href="#contribution"></a>
</h2>
<p>If you encounter any bugs or have suggestions for improvements, please feel free to open an <a href="https://github.com/digitalcytometry/spatialecotyper/issues" class="external-link">issue</a> or submit a pull request. Your feedback and contributions help us make the tool better for everyone.</p>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>Please see the <a href="LICENSE.html" target="_blank">LICENSE</a> file.</p>
</div>
<div class="section level2">
<h2 id="authors">Authors<a class="anchor" aria-label="anchor" href="#authors"></a>
</h2>
<p>Spatial EcoTyper was developed in the <a href="https://anlab.stanford.edu/" target="_blank" class="external-link">Newman Lab</a> by Wubing Zhang.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use Spatial EcoTyper, please cite:</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/digitalcytometry/spatialecotyper" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/digitalcytometry/spatialecotyper/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>STANFORD NON-COMMERCIAL SOFTWARE <a href="LICENSE-text.html">LICENSE</a> AGREEMENT</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SpatialEcoTyper</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2233-299X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
