<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Discovery of Spatial Ecotypes from Multiple Samples • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Discovery of Spatial Ecotypes from Multiple Samples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Spatial.html">Recovery of Spatial Ecotypes from Spatial Transcriptomics Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Discovery of Spatial Ecotypes from Multiple Samples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/vignettes/Integration.Rmd" class="external-link"><code>vignettes/Integration.Rmd</code></a></small>
      <div class="d-none name"><code>Integration.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial, we will illustrate how to identify conserved
spatial ecotypes (SEs) across multiple samples using the <a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a>
function. Each sample represents a single-cell spatial transcriptomics
data, including a gene expression profile and associated single-cell
metadata.</p>
<p>We will be analyzing single-cell spatial transcriptomics data from
cancer samples provided by <a href="https://vizgen.com/data-release-program/" class="external-link">Vizgen’s MERSCOPE FFPE
Human Immuno-oncology</a>. For efficiency, we have selected a subset
regions from a melanoma sample and a colon cancer sample for the
integrative analysis. You can download the gene expression profiles and
single-cell metadata <a href="https://drive.google.com/open?id=1CYncsGZ4aTaIRGkR4rVAmbOGskVyoTEQ&amp;usp=drive_fs" class="external-link"><code>here</code></a>.</p>
<p>The melanoma sample includes spatial expression data for 500 genes
across 27,907 cells, while the colon cancer sample contains data for
38,080 cells. In both samples, cells were categorized into ten distinct
cell types: B cells, CD4+ T cells, CD8+ T cells, NK cells, plasma cells,
macrophages, dendritic cells (DC), fibroblasts, endothelial cells, and
the cancer cell of origin (CCOs). Due to their high heterogeneity, CCOs
were excluded from this demonstration.</p>
<p>All cells were also annotated into four spatial regions, including
tumor, inner margin, outer margin, and stroma. The tumor and stroma
regions were defined based on the density of cancer cells. The inner and
outer margins were delineated as regions extending 250 μm inside and
outside the tumor boundaries, respectively. In addition, we assessed
distance of each single cell to tumor/stroma interface by computing the
shortest Euclidean distance to tumor regions (or stromal regions) for
cells localized in stromal region (or tumor regions). A positive
distance is used for cells localized in tumor region, and a negative
distance is used for cells localized in stroma region.</p>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p><strong>Spatial EcoTyper</strong> analysis requires two input data
for each sample:</p>
<ul>
<li>gene expression matrix: rows represent gene names and columns
represent cell IDs</li>
<li>meta data: a data frame with at least three columns, including “X”
(X-coordinate), “Y” (Y-coordinate), and “CellType” (the cell type
annotation). The row names of the meta data should match the column
names (cell IDs) in the expression matrix.</li>
</ul>
<details open><summary><strong>Text files as input</strong>
</summary><p>First, download the demo data from Google Drive</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"1CoQmU3u8MoVC8RbLUvTDQmOuJJ703HHB"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"1CgUOQKrWY_TG61o5aw7J9LZzE20D6NuI"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"1ChwONUjr_yoURodnkDBj68ZUbjdHtmP6"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"1CipRjjD7cqzqKO0Yf4LUdsEw1XDzP6JS"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_scmeta.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Then, load text files into R. You can use <code>read.table</code> for
small files or <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> for larger files.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell metadata</span></span>
<span><span class="co"># Row names should be cell ids. Required columns: X, Y, CellType; </span></span>
<span><span class="co"># Recommend a 'Region' column if pathologist region annotations are available</span></span>
<span><span class="va">scmeta1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>,</span>
<span>                      sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scmeta2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanColonCancerPatient2_subset_scmeta.tsv"</span>,</span>
<span>                      sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta1</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span>, <span class="st">"Region"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                         X         Y   CellType Region</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast Stroma</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta2</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span>, <span class="st">"Region"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                            X         Y   CellType       Region</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1085 5057.912 -1721.871 Macrophage Inner margin</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1205 5035.339 -1783.352 Macrophage Inner margin</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1312 5012.093 -1908.390 Macrophage Inner margin</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1625 5073.743 -1861.740 Macrophage Inner margin</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1629 5030.235 -1867.441 Macrophage Inner margin</span></span>
<span><span class="co">## HumanColonCancerPatient2__cell_1639 5020.448 -1882.751 Macrophage Inner margin</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell gene expression data. Rows represent gene names and columns represent cell IDs</span></span>
<span><span class="va">scdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>,</span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata1</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co"># Setting the first column as row names</span></span>
<span><span class="va">scdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># Dropping first column</span></span>
<span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanColonCancerPatient2_subset_counts.tsv.gz"</span>,</span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata2</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># Setting the first column as row names</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># Dropping first column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657</span></span>
<span><span class="co">## PDK4                                    0                                1</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     1                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660</span></span>
<span><span class="co">## PDK4                                    1                                0</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     0                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3661</span></span>
<span><span class="co">## PDK4                                    0</span></span>
<span><span class="co">## TNFRSF17                                0</span></span>
<span><span class="co">## ICAM3                                   0</span></span>
<span><span class="co">## FAP                                     0</span></span>
<span><span class="co">## GZMB                                    0</span></span>
<span><span class="co">## TSC2                                    0</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         HumanColonCancerPatient2__cell_1085 HumanColonCancerPatient2__cell_1205</span></span>
<span><span class="co">## PDK4                                      0                                   0</span></span>
<span><span class="co">## CCL26                                     0                                   0</span></span>
<span><span class="co">## CX3CL1                                    0                                   0</span></span>
<span><span class="co">## PGLYRP1                                   0                                   0</span></span>
<span><span class="co">## CD4                                       0                                   0</span></span>
<span><span class="co">## SNAI2                                     0                                   0</span></span>
<span><span class="co">##         HumanColonCancerPatient2__cell_1312 HumanColonCancerPatient2__cell_1625</span></span>
<span><span class="co">## PDK4                                      1                                   0</span></span>
<span><span class="co">## CCL26                                     0                                   0</span></span>
<span><span class="co">## CX3CL1                                    0                                   0</span></span>
<span><span class="co">## PGLYRP1                                   0                                   0</span></span>
<span><span class="co">## CD4                                       0                                   0</span></span>
<span><span class="co">## SNAI2                                     0                                   0</span></span>
<span><span class="co">##         HumanColonCancerPatient2__cell_1629</span></span>
<span><span class="co">## PDK4                                      0</span></span>
<span><span class="co">## CCL26                                     0</span></span>
<span><span class="co">## CX3CL1                                    0</span></span>
<span><span class="co">## PGLYRP1                                   0</span></span>
<span><span class="co">## CD4                                       0</span></span>
<span><span class="co">## SNAI2                                     0</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure row names of the meta data match column names of the gene expression matrix</span></span>
<span><span class="va">scdata1</span> <span class="op">&lt;-</span> <span class="va">scdata1</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="va">scdata2</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</details><details><summary><strong>Sparse matrix as input</strong>
</summary><p>Mtx files can be loaded into R using the <a href="https://satijalab.org/seurat/reference/readmtx" class="external-link"><code>ReadMtx</code>
function</a> from the Seurat package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13M3xhRxp0xK9gf5F4DE9idSBFqVQIXDT"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.mtx.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"136feRaFjMtNvduLTm5xqa3WhyyoG4Xzo"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_cells.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13QprWzJhzzUy_w3XSrjlt9pjf2n-G7HV"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_genes.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17fH9BAAugYi1FMLrMuTxojtOoklFBB-K"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_counts.mtx.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17a1f1VjxJSje_uyPt6zA97zTps9ko6rk"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_cells.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17fH0jE5b2YqJ5fpQtYil27k-7YhzArD6"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_genes.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span>mtx <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_counts.mtx.gz"</span>, cells <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_cells.tsv.gz"</span>, features <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_genes.tsv.gz"</span>, feature.column <span class="op">=</span> <span class="fl">1</span>, cell.column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span>mtx <span class="op">=</span> <span class="st">"HumanColonCancerPatient2_subset_counts.mtx.gz"</span>, cells <span class="op">=</span> <span class="st">"HumanColonCancerPatient2_subset_cells.tsv.gz"</span>, features <span class="op">=</span> <span class="st">"HumanColonCancerPatient2_subset_genes.tsv.gz"</span>, feature.column <span class="op">=</span> <span class="fl">1</span>, cell.column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="data-normalization">Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h2>
<p>Gene expression data should be normalized for the
<code>SpatialEcoTyper</code> analysis. This can be achieved using either
<a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">NormalizeData</a>
or <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2" class="external-link">SCTransform</a>.
Alternatively, you may skip this step and allow the <a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a>
function to handle normalization by specifying the desired method (e.g.,
SCT) in the <code>normalization.method</code> argument.</p>
<details><summary><strong>Using <code>SCTransform</code> for the normalization:</strong>
</summary><p>For faster computation, it is recommended to install the
<code>glmGamPoi</code> package for the <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2" class="external-link">SCTransform</a>
normalization.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Install the glmGamPoi package</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"glmGamPoi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Data normalization</span></span>
<span><span class="va">tmpobj1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>clip.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">tmpobj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>clip.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the normalized gene expression data</span></span>
<span><span class="va">seurat_version</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">seurat_version</span><span class="op">&lt;</span><span class="fl">5</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">normdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">tmpobj1</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span>  <span class="va">normdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">tmpobj2</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="va">normdata1</span> <span class="op">&lt;-</span> <span class="va">tmpobj1</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span></span>
<span>  <span class="va">normdata2</span> <span class="op">&lt;-</span> <span class="va">tmpobj2</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="op">}</span></span></code></pre></div>
</details><details><summary><strong>Using <code>NormalizeData</code> for the normalization:</strong>
</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata1</span><span class="op">)</span></span>
<span><span class="va">normdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="preview-of-the-samples">Preview of the samples<a class="anchor" aria-label="anchor" href="#preview-of-the-samples"></a>
</h2>
<p>The <a href="../reference/SpatialView.html">SpatialView</a> function
can be used to visualize the spatial distribution of single cells within
the tissues.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the cell type annotations in the tissue</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta1</span>, by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"SKCM"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta2</span>, by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CRC"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/viewct-1.png" width="700"></p>
<p>The <a href="../reference/SpatialView.html">SpatialView</a> function
can also be used to visualize spatial landscape of these regions in the
samples.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the regions in the tissue</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta1</span>, by <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"SKCM"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta2</span>, by <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CRC"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/viewregion-1.png" width="700"></p>
<p>You can also visualize continuous characteristics, such as the
minimum distance of each single cell to the tumor/stroma margin, using
the <a href="../reference/SpatialView.html">SpatialView</a> function.
Here, positive distances indicate cells located within the tumor region,
while negative distances denote cells within the stroma.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the distance to tumor margin</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta1</span>, by <span class="op">=</span> <span class="st">"Dist2Interface"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"SKCM"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#5e3c99"</span>, high <span class="op">=</span> <span class="st">"#e66101"</span>, mid <span class="op">=</span> <span class="st">"#d9d9d9"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta2</span>, by <span class="op">=</span> <span class="st">"Dist2Interface"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CRC"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#5e3c99"</span>, high <span class="op">=</span> <span class="st">"#e66101"</span>, mid <span class="op">=</span> <span class="st">"#d9d9d9"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Distance to\ntumor margin"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/viewdist-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Visualize gene expression in the tissue</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta1</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">Expression</span> <span class="op">&lt;-</span> <span class="va">normdata1</span><span class="op">[</span><span class="st">"STAT1"</span>, <span class="op">]</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">gg</span>, by <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"STAT1"</span><span class="op">)</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta2</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">Expression</span> <span class="op">&lt;-</span> <span class="va">normdata2</span><span class="op">[</span><span class="st">"STAT1"</span>, <span class="op">]</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">gg</span>, by <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"STAT1"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/viewgene-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="spatial-ecotyper-analysis-across-multiple-samples">Spatial EcoTyper analysis across multiple samples<a class="anchor" aria-label="anchor" href="#spatial-ecotyper-analysis-across-multiple-samples"></a>
</h2>
<p>The <a href="../reference/SpatialView.html">MultiSpatialEcoTyper</a>
function enables the integration of single-cell spatial transcriptomics
data across multiple samples, facilitating the identification of SEs
shared across samples or cancer types. The analysis is performed in
three stages:</p>
<ul>
<li>
<strong>Independent Spatial EcoTyper Analysis</strong>: <a href="../reference/SpatialEcoTyper.html">Spatial EcoTyper</a> is run on
each sample independently to identify spatial clusters and construct
cell-type-specific gene expression profiles for all spatial
clusters.</li>
<li>
<strong>Integration</strong>: Spatial
clusters from different samples are integrated to learn a unified
embedding.</li>
<li>
<strong>Clustering</strong>: Clustering is applied to the unified embedding to identify conserved SEs
across samples.</li>
</ul>
<details open><summary><strong>Key arguments for <code>MultiSpatialEcoTyper</code></strong>
</summary><ul>
<li>
<code>data_list</code> A named list of expression matrices where
each matrix represents gene expression data for a sample. The columns of
each matrix correspond to cells, and the rows correspond to genes. The
list names should match sample names; otherwise, samples will be named
as ‘Sample1’, ‘Sample2’, etc.</li>
<li>
<code>metadata_list</code> A named list of metadata data frames,
where each data frame contains metadata corresponding to the cells in
the expression matrices. Each metadata data frame should include at
least three columns (X, Y, CellType), and the row names should match the
cell IDs (column names) in the expression matrix.</li>
<li>
<code>outdir</code> Directory where the results will be saved.
Defaults to the current directory with a subdirectory named
“SpatialEcoTyper_results_” followed by the current date.</li>
<li>
<code>normalization.method</code> Method for normalizing the
expression data. Options include “None” (default), “SCT”, or other
methods compatible with Seurat’s <code>NormalizeData</code>
function.</li>
<li>
<code>nmf_ranks</code> Integer or vector specifying the number of
clusters (10 by default). If a vector is provided, the function tests
all ranks and selects the optimal rank for NMF, though this may increase
computation time.</li>
<li>
<code>radius</code> Numeric specifying the radius (in the same units
as spatial coordinates) for defining spatial neighborhoods around each
cell. Default is 50.</li>
<li>
<code>minibatch</code> Integer specifying the number of columns
(default: 5000) to process in each minibatch in the SNF analysis. This
option splits the matrix into smaller chunks, thus reducing memory
usage.</li>
<li>
<code>ncores</code> Integer specifying the number of cores for
parallel processing (default: 1).</li>
</ul>
<p>You can type <code><a href="../reference/MultiSpatialEcoTyper.html">?MultiSpatialEcoTyper</a></code> to visualize the full
manual.</p>
</details><p><strong>Important note</strong>: In the examples below, we will use
10 clusters and 5 runs per rank for demonstration purposes. To select
the optimal number of clusters and obtain robust results, you can use
the <code>nmf_ranks</code> argument with a vector of integers to test
multiple ranks for NMF analysis. You can also use the <a href="../reference/nmfClustering.html">nmfClustering</a> function for
the rank selection, as outlined in the next section
(<code>NMF clustering for SpatialEcoTyper</code>).</p>
<div class="section level3">
<h3 id="preparing-data">Preparing data<a class="anchor" aria-label="anchor" href="#preparing-data"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SKCM <span class="op">=</span> <span class="va">normdata1</span>, CRC <span class="op">=</span> <span class="va">normdata2</span><span class="op">)</span></span>
<span><span class="va">metadata_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SKCM <span class="op">=</span> <span class="va">scmeta1</span>, CRC <span class="op">=</span> <span class="va">scmeta2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pathologist-region-annotation-unavailable">Pathologist region annotation unavailable<a class="anchor" aria-label="anchor" href="#pathologist-region-annotation-unavailable"></a>
</h3>
<details open><summary><code>MultiSpatialEcoTyper</code> analysis (without region annotation)
</summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a></span><span class="op">(</span><span class="va">data_list</span>, <span class="va">metadata_list</span>, </span>
<span>                     normalization.method <span class="op">=</span> <span class="st">"None"</span>, <span class="co"># Normalization method</span></span>
<span>                     nmf_ranks <span class="op">=</span> <span class="fl">10</span>,  <span class="co"># Number of clusters</span></span>
<span>                     nrun.per.rank <span class="op">=</span> <span class="fl">5</span>,  <span class="co"># Recommend 30 or higher for robust results</span></span>
<span>                     ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># This demo takes ~3 minutes on a macOS with an Apple M1 Pro chip and 16 GB memory.</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="pathologist-region-annotation-available">Pathologist region annotation available<a class="anchor" aria-label="anchor" href="#pathologist-region-annotation-available"></a>
</h3>
<p>If region annotations are available, we recommend to include the
region annotations in all metadata and specify the <code>Region</code>
in the <a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a>
function, which could improve integration accuracy.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Annotation of tumor and stroma regions will be considered for the analysis.</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Dist2Interface</span><span class="op">&lt;</span><span class="fl">0</span>, <span class="st">"Stroma"</span>, <span class="st">"Tumor"</span><span class="op">)</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Dist2Interface</span><span class="op">&lt;</span><span class="fl">0</span>, <span class="st">"Stroma"</span>, <span class="st">"Tumor"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a></span><span class="op">(</span><span class="va">data_list</span>, <span class="va">metadata_list</span>, </span>
<span>                     normalization.method <span class="op">=</span> <span class="st">"None"</span>, <span class="co"># Normalization method</span></span>
<span>                     min.cts.per.region <span class="op">=</span> <span class="fl">2</span>, <span class="co"># At least two cell types in each microregion</span></span>
<span>                     nmf_ranks <span class="op">=</span> <span class="fl">10</span>,  <span class="co"># Number of clusters</span></span>
<span>                     nrun.per.rank <span class="op">=</span> <span class="fl">5</span>,  <span class="co"># Recommend 30 or higher for robust results</span></span>
<span>                     Region <span class="op">=</span> <span class="st">"Region2"</span>, <span class="co"># Use pathologist annotations if available</span></span>
<span>                     ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">## This demo takes ~3 minutes to complete on macOS with an Apple M1 Pro chip and 16 GB memory.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pathologist-region-annotation-available-2">Pathologist region annotation available (2)<a class="anchor" aria-label="anchor" href="#pathologist-region-annotation-available-2"></a>
</h3>
<p>You can specify the <code>nmf_ranks</code> argument with a vector of
integers to test multiple ranks, enabling the selection of the optimal
number of clusters.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Annotation of tumor and stroma regions will be considered for the analysis.</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">metadata_list</span><span class="op">$</span><span class="va">SKCM</span><span class="op">$</span><span class="va">Dist2Interface</span><span class="op">&lt;</span><span class="fl">0</span>, <span class="st">"Stroma"</span>, <span class="st">"Tumor"</span><span class="op">)</span></span>
<span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Region2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">metadata_list</span><span class="op">$</span><span class="va">CRC</span><span class="op">$</span><span class="va">Dist2Interface</span><span class="op">&lt;</span><span class="fl">0</span>, <span class="st">"Stroma"</span>, <span class="st">"Tumor"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/MultiSpatialEcoTyper.html">MultiSpatialEcoTyper</a></span><span class="op">(</span><span class="va">data_list</span>, <span class="va">metadata_list</span>, </span>
<span>                     normalization.method <span class="op">=</span> <span class="st">"None"</span>, <span class="co"># Normalization method</span></span>
<span>                     nmf_ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">15</span>,<span class="fl">2</span><span class="op">)</span>,  <span class="co"># Number of clusters</span></span>
<span>                     nrun.per.rank <span class="op">=</span> <span class="fl">30</span>,  <span class="co"># 30 runs per rank for robust results</span></span>
<span>                     Region <span class="op">=</span> <span class="st">"Region2"</span>, <span class="co"># Use pathologist annotations</span></span>
<span>                     ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">## This demo takes ~16 minutes to complete on macOS with an Apple M1 Pro chip and 16 GB memory.</span></span></code></pre></div>
<details><summary><strong>Using pre-existing Spatial EcoTyper results</strong>
</summary><p>If you’ve already run <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a> for each
sample, you can integrate the results directly using <a href="../reference/IntegrateSpatialEcoTyper.html">IntegrateSpatialEcoTyper</a>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"126OSOk8sAaxhJID1D3-75bI9S-to9Lkg"</span><span class="op">)</span>, <span class="st">"CRC_SpatialEcoTyper_results.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"11y0lcpawQAZAvDCnt3tBQsF90Wet0lKe"</span><span class="op">)</span>, <span class="st">"SKCM_SpatialEcoTyper_results.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SKCM <span class="op">=</span> <span class="va">normdata1</span>, CRC <span class="op">=</span> <span class="va">normdata2</span><span class="op">)</span></span>
<span><span class="va">SpatialEcoTyper_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SKCM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SKCM_SpatialEcoTyper_results.rds"</span><span class="op">)</span>,</span>
<span>                             CRC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"CRC_SpatialEcoTyper_results.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/IntegrateSpatialEcoTyper.html">IntegrateSpatialEcoTyper</a></span><span class="op">(</span><span class="va">SpatialEcoTyper_list</span>, <span class="va">data_list</span>, </span>
<span>                         outdir <span class="op">=</span> <span class="st">"SpatialEcoTyper_results/"</span>,</span>
<span>                         normalization.method <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>                         nmf_ranks <span class="op">=</span> <span class="fl">10</span>, <span class="co"># Number of clusters</span></span>
<span>                         nrun.per.rank <span class="op">=</span> <span class="fl">5</span>,  <span class="co"># recommend 30 or higher for robust results</span></span>
<span>                         Region <span class="op">=</span> <span class="st">"Region2"</span>, <span class="co"># Use pathologist annotations if available</span></span>
<span>                         ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># This demo takes ~3 minutes on a macOS with an Apple M1 Pro chip and 16 GB memory.</span></span></code></pre></div>
</details><details><summary><strong>Optimizing memory usage</strong>
</summary><p>For real-world single-cell ST datasets with over 100,000 cells, the
analysis can be both time- and memory-intensive. To speed up the
process, you can increase the number of cores used
(<code>ncores</code>), but this will also increase memory
consumption.</p>
<p>If you have limited computational memory, there are several
strategies to reduce usage. One approach is to decrease the
<code>minibatch</code> size and reduce the number of cores
(<code>ncores</code>) allocated. However, for larger datasets, the
minimum memory requirement may remain high. Another option is to
increase the <code>grid.size</code>, which reduces the number of spatial
microregions. This can significantly decrease memory usage by
downsampling the ST data, potentially excluding many cells from the
analysis though.</p>
</details><details><summary><strong>Interested in multicellular communities associated with specific
cell types?</strong>
</summary><p>If you’re interested in studying multicellular communities associated
with specific cell types, you can use the
<code>filter.region.by.celltypes</code> argument to focus on
microregions containing at least one cell from the specified cell types.
An example is available in <a href="./SingleSample.html">Discovery of
Spatial Ecotypes from a Single Sample</a>.</p>
</details><details><summary><strong>Interested in regions with multiple cell types?</strong>
</summary><p>If you’re interested in regions composed of multiple cell types, you
can use the <code>min.cts.per.region</code> argument to specify the
minimum number of distinct cell types required for each microregion to
be included in the analysis.</p>
</details>
</div>
</div>
<div class="section level2">
<h2 id="output-results">Output results<a class="anchor" aria-label="anchor" href="#output-results"></a>
</h2>
<p>All results will be saved in the specified directory
(<code>outdir</code>) with a subdirectory named
<code>SpatialEcoTyper_results_</code> followed by the current date. The
output files include:</p>
<ul>
<li>
<strong><code>(SampleName)_SpatialEcoTyper_results.rds</code></strong>
<code>SpatialEcoTyper</code> result for each sample.</li>
<li>
<strong>MultiSE_integrated_final.rds</strong> A matrix representing
the fused similarity matrix of spatial clusters across all samples.</li>
<li>
<strong>MultiSE_integrated_final_hmap.pdf</strong> A heatmap
visualizing the fused similarity matrix of spatial clusters, grouped by
samples and SEs.</li>
<li>
<strong>MultiSE_metadata_final.rds</strong> Single-cell metadata
with an ‘SE’ column annotating the discovered SEs, saved as an RDS
file.</li>
<li>
<strong>MultiSE_metadata_final.tsv</strong> The same as
<code>MultiSE_metadata_final.rds</code>, but saved in TSV format.</li>
<li>
<strong>SpatialView_SEs_by_Sample.pdf</strong> A figure showing the
spatial landscape of SEs across samples.</li>
<li>
<strong>BarView_SEs_Sample_Frac.pdf</strong> A bar plot showing the
fraction of cells from each sample represented within each SE.</li>
<li>
<strong>BarView_SEs_Region_Frac_Avg.pdf</strong> A bar plot showing
the average fraction of regions represented within each SE.</li>
<li>
<strong>BarView_SEs_CellType_Frac_Avg.pdf</strong> A bar plot
showing the average cell type composition within each SE.</li>
<li>
<strong>MultiSE_NMF_results.rds</strong> The NMF result derived from
the <code>nmfClustering</code> function. If a single rank was provided,
it contains an <code>NMFfitX1</code> object . If multiple ranks were
provided, it is a <code>list</code> containing the optimal number of
communities (<code>bestK</code>), a list of <code>NMFfitX1</code>
objects (<code>NMFfits</code>), and a <code>ggplot</code> object
(<code>p</code>) displaying the cophenetic coefficient for different
cluster numbers.</li>
<li>
<strong>MultiSE_NMF_Cophenetic_Dynamics.pdf</strong> A figure
showing the cophenetic coefficients for different numbers of clusters,
only available when multiple ranks were provided.</li>
</ul>
<p>The results for the demo data can be downloaded from <a href="https://drive.google.com/open?id=1H4S71HENMoig03O3FdeLwnwk2aja-Wtv&amp;usp=drive_fs" class="external-link"><code>MultiSpatialEcoTyper_results</code></a>.</p>
</div>
<div class="section level2">
<h2 id="nmf-clustering">NMF clustering<a class="anchor" aria-label="anchor" href="#nmf-clustering"></a>
</h2>
<p><strong>Spatial EcoTyper</strong> uses NMF to group spatial clusters
from multiple samples into conserved SEs. To determine the optimal
number of clusters, or rank, we computed cophenetic coefficient, which
quantifies the classification stability for a given rank (i.e., the
number of clusters) and ranges from 0 to 1, with 1 indicating maximally
stability. Typically, the rank at which the cophenetic coefficient
begins to decrease is chosen. However, this method can be challenging
when the cophenetic coefficient exhibits a multi-modal shape across
different ranks. In such cases, for <strong>Spatial EcoTyper</strong>
analysis, we recommend selecting the number of SEs after which the
coefficient drops below 0.95 by default.</p>
<div class="section level3">
<h3 id="selecting-the-optimal-rank">Selecting the optimal rank<a class="anchor" aria-label="anchor" href="#selecting-the-optimal-rank"></a>
</h3>
<p>To determine the optimal rank, you can use the <a href="../reference/nmfClustering.html">nmfClustering</a> function, which
tests multiple ranks and computes the cophenetic coefficient for each.
This analysis requires the fused similarity network matrix
(<code>MultiSE_integrated_final.rds</code>) generated from previous
step.</p>
<p>Load the files into R:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Fused similarity matrix of spatial clusters across all samples</span></span>
<span><span class="va">outdir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SpatialEcoTyper_results_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">integrated</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"MultiSE_integrated_final.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">integrated</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 376 376</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Single-cell metadata with an 'SE' column annotating the discovered SEs</span></span>
<span><span class="va">finalmeta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"MultiSE_metadata_final.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finalmeta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span>, <span class="st">"Region"</span>, <span class="st">"InitSE"</span>, <span class="st">"SE"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  Sample        X         Y   CellType Region</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655   SKCM 1894.706 -6367.766 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657   SKCM 1942.480 -6369.602 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658   SKCM 1963.007 -6374.026 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660   SKCM 1981.600 -6372.266 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661   SKCM 1742.939 -6374.851 Fibroblast Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663   SKCM 1921.683 -6383.309 Fibroblast Stroma</span></span>
<span><span class="co">##                                          InitSE  SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 SKCM..InitSE77 SE3</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 SKCM..InitSE63 SE4</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 SKCM..InitSE63 SE4</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 SKCM..InitSE63 SE4</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 SKCM..InitSE60 SE3</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 SKCM..InitSE77 SE3</span></span></code></pre>
<p>Next, test multiple ranks to find the optimal number of SEs. The
running time for this step depends on the data size, number of ranks to
test, the number of runs per rank, and the number of cores for parallel
analysis. For demonstration, here we test six different ranks ranging
from 4 to 15 with 5 runs per rank.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmf_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nmfClustering.html">nmfClustering</a></span><span class="op">(</span><span class="va">integrated</span>, ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">15</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                         nrun.per.rank <span class="op">=</span> <span class="fl">5</span>, <span class="co"># Recommend 30 or higher</span></span>
<span>                         min.coph <span class="op">=</span> <span class="fl">0.99</span>, <span class="co"># minimum cophenetic coefficient threshold</span></span>
<span>                         ncores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                         seed <span class="op">=</span> <span class="fl">2024</span><span class="op">)</span></span>
<span><span class="co">## This process takes ~3 minutes on a macOS with an Apple M1 Pro chip and 16 GB of memory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"The selected rank is "</span>, <span class="va">nmf_res</span><span class="op">$</span><span class="va">bestK</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "The selected rank is 4"</span></span></code></pre>
<p>To better understand how the cophenetic coefficient changes with
different ranks, you can visualize the results.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nmf_res</span><span class="op">$</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/coph-1.png" width="700"></p>
<p>Once the optimal rank is determined, you can use it to generate the
final SE annotations.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ses</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">nmf_res</span><span class="op">$</span><span class="va">NMFfits</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"K."</span>, <span class="va">nmf_res</span><span class="op">$</span><span class="va">bestK</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ses</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   SKCM..InitSE1  SKCM..InitSE74 SKCM..InitSE110  SKCM..InitSE16 SKCM..InitSE111 </span></span>
<span><span class="co">##               1               2               2               2               2 </span></span>
<span><span class="co">##  SKCM..InitSE24 </span></span>
<span><span class="co">##               2 </span></span>
<span><span class="co">## Levels: 1 2 3 4</span></span></code></pre>
<p>You can add the newly defined SEs into the meta data</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finalmeta</span><span class="op">$</span><span class="va">SE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SE"</span>, <span class="va">ses</span><span class="op">[</span><span class="va">finalmeta</span><span class="op">$</span><span class="va">InitSE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finalmeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                               CID        X</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3655 1894.706</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 HumanMelanomaPatient1__cell_3657 1942.480</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3658 1963.007</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 HumanMelanomaPatient1__cell_3660 1981.600</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 HumanMelanomaPatient1__cell_3661 1742.939</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 HumanMelanomaPatient1__cell_3663 1921.683</span></span>
<span><span class="co">##                                          Y   CellType CellTypeName Region</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 -6367.766 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 -6369.602 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 -6374.026 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 -6372.266 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 -6374.851 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 -6383.309 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">##                                  Dist2Interface Region2         InitSE Sample</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655      -883.1752  Stroma SKCM..InitSE77   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657      -894.8463  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658      -904.1115  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660      -907.8909  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661      -874.2712  Stroma SKCM..InitSE60   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663      -903.6559  Stroma SKCM..InitSE77   SKCM</span></span>
<span><span class="co">##                                   SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 SE1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-a-different-rank">Using a different rank<a class="anchor" aria-label="anchor" href="#using-a-different-rank"></a>
</h3>
<p>If you simply want to experiment with a different rank, you can use
the <a href="../reference/nmfClustering.html">nmfClustering</a>
function. When a single rank is specified, it directly returns an NMFfit
object for predicting SE grouping, as shown below:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmf_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/nmfClustering.html">nmfClustering</a></span><span class="op">(</span><span class="va">integrated</span>, ranks <span class="op">=</span> <span class="fl">10</span>, nrun.per.rank <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                         seed <span class="op">=</span> <span class="fl">1</span>, ncores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ses</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">nmf_res</span><span class="op">)</span></span>
<span><span class="co">## This process takes ~6 minutes on a macOS with an Apple M1 Pro chip and 16 GB of memory</span></span></code></pre></div>
<p>You can add the newly defined SEs into the meta data</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finalmeta</span><span class="op">$</span><span class="va">SE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SE"</span>, <span class="va">ses</span><span class="op">[</span><span class="va">finalmeta</span><span class="op">$</span><span class="va">InitSE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">finalmeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                               CID        X</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3655 1894.706</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 HumanMelanomaPatient1__cell_3657 1942.480</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3658 1963.007</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 HumanMelanomaPatient1__cell_3660 1981.600</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 HumanMelanomaPatient1__cell_3661 1742.939</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 HumanMelanomaPatient1__cell_3663 1921.683</span></span>
<span><span class="co">##                                          Y   CellType CellTypeName Region</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 -6367.766 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 -6369.602 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 -6374.026 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 -6372.266 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 -6374.851 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 -6383.309 Fibroblast  Fibroblasts Stroma</span></span>
<span><span class="co">##                                  Dist2Interface Region2         InitSE Sample</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655      -883.1752  Stroma SKCM..InitSE77   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657      -894.8463  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658      -904.1115  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660      -907.8909  Stroma SKCM..InitSE63   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661      -874.2712  Stroma SKCM..InitSE60   SKCM</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663      -903.6559  Stroma SKCM..InitSE77   SKCM</span></span>
<span><span class="co">##                                    SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655  SE4</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 SE10</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 SE10</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 SE10</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661  SE4</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663  SE4</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-ses">Visualizing SEs<a class="anchor" aria-label="anchor" href="#visualizing-ses"></a>
</h2>
<div class="section level3">
<h3 id="visualizing-the-integrated-similarity-matrix">Visualizing the integrated similarity matrix<a class="anchor" aria-label="anchor" href="#visualizing-the-integrated-similarity-matrix"></a>
</h3>
<p>After reordering spatial clusters based on the newly identified SEs,
you can re-draw the heatmap to show the fused similarity matrix of
spatial clusters across samples.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ords</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ses</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">ses</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">integrated</span> <span class="op">=</span> <span class="va">integrated</span><span class="op">[</span><span class="va">ords</span>, <span class="va">ords</span><span class="op">]</span> <span class="co"># reorder the cells</span></span>
<span><span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\.\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">integrated</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  SE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SE"</span>, <span class="va">ses</span><span class="op">[</span><span class="va">ords</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                  row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">integrated</span><span class="op">)</span><span class="op">)</span> <span class="co"># cell groups</span></span>
<span><span class="va">SE_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getColors.html">getColors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span>, palette <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># colors for SEs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">SE_cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="va">sample_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getColors.html">getColors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span><span class="op">)</span>, palette <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># colors for samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sample_cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## draw heatmap</span></span>
<span><span class="fu"><a href="../reference/HeatmapView.html">HeatmapView</a></span><span class="op">(</span><span class="va">integrated</span>, show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            top_ann <span class="op">=</span> <span class="va">ann</span>, top_ann_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Sample <span class="op">=</span> <span class="va">sample_cols</span>, SE <span class="op">=</span> <span class="va">SE_cols</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawRectangleAnnotation.html">drawRectangleAnnotation</a></span><span class="op">(</span><span class="va">ann</span><span class="op">$</span><span class="va">SE</span>, <span class="va">ann</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/newsehmap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-ses-in-the-tissue">Visualizing SEs in the tissue<a class="anchor" aria-label="anchor" href="#visualizing-ses-in-the-tissue"></a>
</h3>
<p>To examine the spatial distribution of the identified SEs, you can
use the <a href="../reference/SpatialView.html">SpatialView</a> function
visualize the SEs in the tissues.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">finalmeta</span>, by <span class="op">=</span> <span class="st">"SE"</span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Sample</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/spatialmapse-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="association-between-ses-and-pre-annotated-regions">Association between SEs and pre-annotated regions<a class="anchor" aria-label="anchor" href="#association-between-ses-and-pre-annotated-regions"></a>
</h3>
<p>To examine the enrichment of SEs in pre-annotated regions, you can
visualize the cell fractions from each region, for each SE.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Region</span>, <span class="va">Sample</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Sample</span>, <span class="va">SE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Frac <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">## cell type fractions within each sample</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Region</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>Frac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Frac</span><span class="op">)</span><span class="op">)</span> <span class="co">## average cell type fractions across all samples</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Frac</span>, fill <span class="op">=</span> <span class="va">Region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Fraction"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/bar_region1-1.png" width="700"></p>
<details><summary><strong>Association between SEs and pre-annotated regions within each
sample</strong>
</summary><div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Region</span>, <span class="va">Sample</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">n</span>, fill <span class="op">=</span> <span class="va">Region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Sample</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Fraction of regions"</span><span class="op">)</span></span></code></pre></div>
<img src="Integration_files/figure-html/bar_region2-1.png" width="700"></details>
</div>
<div class="section level3">
<h3 id="distance-of-ses-to-tumorstroma-interface">Distance of SEs to tumor/stroma interface<a class="anchor" aria-label="anchor" href="#distance-of-ses-to-tumorstroma-interface"></a>
</h3>
<p>This box plot visualizes the distribution of distances of SEs to the
tumor/stroma interface. Positive distances indicate cells located within
the tumor region, while negative distances denote cells within the
stroma. The SEs are ordered by their median distance, highlighting their
spatial localization relative to the tumor/stroma interface.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Sample</span>, <span class="va">SE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>Dist2Interface <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Dist2Interface</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Dist2Interface</span><span class="op">)</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">SE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">gg</span><span class="op">$</span><span class="va">SE</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gg</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Dist2Interface</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Distance to tumor/stroma interface (μm)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/box_dist-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="cell-type-composition-of-ses">Cell type composition of SEs<a class="anchor" aria-label="anchor" href="#cell-type-composition-of-ses"></a>
</h3>
<p>We can visualize the cell type composition of SEs in a bar plot. The
cell type fractions can be averaged across the two samples.</p>
<p><strong>Average cell type composition of SEs across all
samples</strong></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">CellType</span>, <span class="va">Sample</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Sample</span>, <span class="va">SE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Frac <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">## cell type fractions within each sample</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">CellType</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>Frac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Frac</span><span class="op">)</span><span class="op">)</span> <span class="co">## average cell type fractions across all samples</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Frac</span>, fill <span class="op">=</span> <span class="va">CellType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell type abundance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/bar_ct2-1.png" width="700"></p>
<details><summary><strong>Cell type composition of SEs within each sample</strong>
</summary><div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">CellType</span>, <span class="va">Sample</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">n</span>, fill <span class="op">=</span> <span class="va">CellType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Sample</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell type abundance"</span><span class="op">)</span></span></code></pre></div>
<img src="Integration_files/figure-html/bar_ct1-1.png" width="700"></details>
</div>
</div>
<div class="section level2">
<h2 id="identification-of-cell-type-specific-se-markers">Identification of cell-type-specific SE markers<a class="anchor" aria-label="anchor" href="#identification-of-cell-type-specific-se-markers"></a>
</h2>
<p>You can identify cell-type-specific SE markers by differential
expression analysis using the <a href="https://github.com/immunogenomics/presto" class="external-link"><code>presto</code>
package</a>. Below is an example of how to identify fibroblast-specific
markers for each SE.</p>
<div class="section level3">
<h3 id="de-analysis-within-each-sample">DE analysis within each sample<a class="anchor" aria-label="anchor" href="#de-analysis-within-each-sample"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"presto"</span><span class="op">)</span></span>
<span><span class="co">## DE analysis within the first sample</span></span>
<span><span class="va">tmpmeta1</span> <span class="op">=</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CellType</span><span class="op">==</span><span class="st">"Fibroblast"</span> <span class="op">&amp;</span> <span class="va">Sample</span><span class="op">==</span><span class="st">"SKCM"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmpdata1</span> <span class="op">=</span> <span class="va">normdata1</span><span class="op">[</span>, <span class="va">tmpmeta1</span><span class="op">$</span><span class="va">CID</span><span class="op">]</span></span>
<span><span class="va">degs1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/presto/man/wilcoxauc.html" class="external-link">wilcoxauc</a></span><span class="op">(</span><span class="va">tmpdata1</span>, <span class="va">tmpmeta1</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="co">## DE analysis within the second sample</span></span>
<span><span class="va">tmpmeta2</span> <span class="op">=</span> <span class="va">finalmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CellType</span><span class="op">==</span><span class="st">"Fibroblast"</span> <span class="op">&amp;</span> <span class="va">Sample</span><span class="op">==</span><span class="st">"CRC"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmpdata2</span> <span class="op">=</span> <span class="va">normdata2</span><span class="op">[</span>, <span class="va">tmpmeta2</span><span class="op">$</span><span class="va">CID</span><span class="op">]</span></span>
<span><span class="va">degs2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/presto/man/wilcoxauc.html" class="external-link">wilcoxauc</a></span><span class="op">(</span><span class="va">tmpdata2</span>, <span class="va">tmpmeta2</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span> <span class="co"># DE analysis</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">degs1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    feature group     avgExpr        logFC statistic       auc         pval</span></span>
<span><span class="co">## 1     PDK4  SE10 0.117555097  0.047787757   4701164 0.5269201 2.300787e-09</span></span>
<span><span class="co">## 2 TNFRSF17  SE10 0.000000000  0.000000000   4460984 0.5000000 1.000000e+00</span></span>
<span><span class="co">## 3    ICAM3  SE10 0.000000000  0.000000000   4460984 0.5000000 1.000000e+00</span></span>
<span><span class="co">## 4      FAP  SE10 0.102200894 -0.037071606   4241882 0.4754425 1.963176e-05</span></span>
<span><span class="co">## 5     GZMB  SE10 0.007441764 -0.008093071   4423277 0.4957737 3.439501e-02</span></span>
<span><span class="co">## 6     TSC2  SE10 0.006910210 -0.002427312   4446112 0.4983331 3.257778e-01</span></span>
<span><span class="co">##           padj     pct_in   pct_out</span></span>
<span><span class="co">## 1 1.030353e-08 14.2638037  9.003215</span></span>
<span><span class="co">## 2 1.000000e+00  0.0000000  0.000000</span></span>
<span><span class="co">## 3 1.000000e+00  0.0000000  0.000000</span></span>
<span><span class="co">## 4 6.684535e-05 13.3435583 18.269512</span></span>
<span><span class="co">## 5 8.335732e-02  1.0736196  1.914645</span></span>
<span><span class="co">## 6 6.484080e-01  0.9969325  1.330020</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">degs2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    feature group    avgExpr       logFC statistic       auc         pval</span></span>
<span><span class="co">## 1     PDK4   SE1 0.07290325 -0.08483371  15082423 0.4580113 3.871812e-25</span></span>
<span><span class="co">## 2   CX3CL1   SE1 0.03282075 -0.01465999  16182732 0.4914246 7.090800e-04</span></span>
<span><span class="co">## 3      CD4   SE1 0.00000000  0.00000000  16465120 0.5000000 1.000000e+00</span></span>
<span><span class="co">## 4    SNAI2   SE1 0.35501044  0.11399360  18215060 0.5531408 6.627100e-26</span></span>
<span><span class="co">## 5 TNFRSF17   SE1 0.00000000  0.00000000  16465120 0.5000000 1.000000e+00</span></span>
<span><span class="co">## 6    ICAM3   SE1 0.00000000  0.00000000  16465120 0.5000000 1.000000e+00</span></span>
<span><span class="co">##           padj    pct_in   pct_out</span></span>
<span><span class="co">## 1 2.380875e-24  8.921162 17.125293</span></span>
<span><span class="co">## 2 1.848993e-03  4.107884  5.818208</span></span>
<span><span class="co">## 3 1.000000e+00  0.000000  0.000000</span></span>
<span><span class="co">## 4 4.333913e-25 36.431535 26.939403</span></span>
<span><span class="co">## 5 1.000000e+00  0.000000  0.000000</span></span>
<span><span class="co">## 6 1.000000e+00  0.000000  0.000000</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="identifying-conserved-markers-across-samples">Identifying conserved markers across samples<a class="anchor" aria-label="anchor" href="#identifying-conserved-markers-across-samples"></a>
</h3>
<p>To identify markers conserved across different samples, we conduct a
meta-analysis by averaging the log2 fold changes (log2FC) from the DE
analyses.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">degs1</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span>, <span class="va">degs2</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature"</span>, <span class="st">"group"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">degs</span><span class="op">$</span><span class="va">AvgLogFC</span> <span class="op">=</span> <span class="op">(</span><span class="va">degs</span><span class="op">$</span><span class="va">logFC.x</span> <span class="op">+</span> <span class="va">degs</span><span class="op">$</span><span class="va">logFC.y</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="va">lfcs</span> <span class="op">=</span> <span class="va">degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="va">feature</span>, names_from <span class="op">=</span> <span class="va">group</span>, </span>
<span>                            values_from <span class="op">=</span> <span class="va">AvgLogFC</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">as.data.frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">lfcs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lfcs</span><span class="op">$</span><span class="va">feature</span></span>
<span><span class="va">lfcs</span> <span class="op">&lt;-</span> <span class="va">lfcs</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lfcs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                SE10         SE3         SE4         SE6         SE7         SE8</span></span>
<span><span class="co">## ACKR3    0.04539229 -0.11089193  0.22761738 -0.13915649 -0.16230522  0.05704506</span></span>
<span><span class="co">## ACTA2    0.64868983  0.19121333 -0.28277467 -0.15495698 -0.13060735  0.26294794</span></span>
<span><span class="co">## ADAMTS4 -0.07880286 -0.03558349 -0.10298639  0.15231638  0.14598004 -0.10437673</span></span>
<span><span class="co">## AKT1    -0.11112575 -0.00885807 -0.19659536  0.11531490  0.32156308 -0.09316424</span></span>
<span><span class="co">## AKT2     0.01227246 -0.01219427  0.01067868 -0.01456746  0.01794543 -0.02832473</span></span>
<span><span class="co">## AKT3     0.05820973 -0.05104454 -0.02226315 -0.01922768  0.02741800 -0.03496944</span></span>
<span><span class="co">##                   SE9</span></span>
<span><span class="co">## ACKR3   -2.297758e-02</span></span>
<span><span class="co">## ACTA2   -1.850714e-01</span></span>
<span><span class="co">## ADAMTS4 -3.370712e-03</span></span>
<span><span class="co">## AKT1    -3.274771e-02</span></span>
<span><span class="co">## AKT2    -2.510557e-05</span></span>
<span><span class="co">## AKT3    -2.496374e-02</span></span></code></pre>
<p>To identify markers that are specific to each SE-associated cell
state, we calculate the difference between the maximum log2FC for each
gene and the second-highest log2FC:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">secondmax</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">lfcs</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">-</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">=</span> <span class="va">lfcs</span> <span class="op">-</span> <span class="va">secondmax</span></span>
<span><span class="co">## Markers are considered specific if they have a positive delta and a log2FC greater than 0.05:</span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="va">delta</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">lfcs</span><span class="op">&gt;</span><span class="fl">0.05</span></span>
<span><span class="va">markers</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lfcs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">[</span>, <span class="va">se</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">gs</span> <span class="op">=</span> <span class="va">gs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">lfcs</span><span class="op">[</span><span class="va">gs</span>, <span class="va">se</span><span class="op">]</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">gs</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lfcs</span><span class="op">)</span></span>
<span><span class="va">markers</span> <span class="op">=</span> <span class="va">markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">]</span></span>
<span><span class="co">## Select the top five markers</span></span>
<span><span class="va">top5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">top5</span></span></code></pre></div>
<pre><code><span><span class="co">## $SE10</span></span>
<span><span class="co">## [1] "ACTA2"  "COL4A1" "MYH11"  "TGFB3"  "TNC"   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE3</span></span>
<span><span class="co">## [1] "FN1"    "CSF1"   "COL5A1"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE4</span></span>
<span><span class="co">## [1] "FOS"     "PLA2G2A" "DUSP1"   "EGR1"    "FBLN1"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE6</span></span>
<span><span class="co">## [1] "HLA-DRA" "BST2"    "CXCL9"   "STAT1"   "SOD2"   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE7</span></span>
<span><span class="co">## [1] "PKM"    "MMP11"  "CTNNB1" "TGFBI"  "YAP1"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE8</span></span>
<span><span class="co">## [1] "SFRP2"   "COL1A1"  "SMOC2"   "COL11A1" "MFAP5"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $SE9</span></span>
<span><span class="co">## [1] "XBP1"  "CXCR4" "CCL2"  "PTPRC" "STAT3"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualizing-top-se-cell-state-markers">Visualizing top SE cell state markers<a class="anchor" aria-label="anchor" href="#visualizing-top-se-cell-state-markers"></a>
</h3>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">lfcs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">top5</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/HeatmapView.html">HeatmapView</a></span><span class="op">(</span><span class="va">gg</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Integration_files/figure-html/hmaptop-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      parallel  stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] tidyr_1.3.1           presto_1.0.0          Rcpp_1.0.13          </span></span>
<span><span class="co">##  [4] pals_1.9              SpatialEcoTyper_1.0.0 RANN_2.6.2           </span></span>
<span><span class="co">##  [7] Matrix_1.7-0          ComplexHeatmap_2.20.0 NMF_0.28             </span></span>
<span><span class="co">## [10] Biobase_2.64.0        BiocGenerics_0.50.0   cluster_2.1.6        </span></span>
<span><span class="co">## [13] rngtools_1.5.2        registry_0.5-1        googledrive_2.1.1    </span></span>
<span><span class="co">## [16] data.table_1.16.0     Seurat_5.1.0          SeuratObject_5.0.2   </span></span>
<span><span class="co">## [19] sp_2.1-4              ggplot2_3.5.1         dplyr_1.1.4          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">##   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">##   [5] R.oo_1.26.0                 polyclip_1.10-7            </span></span>
<span><span class="co">##   [7] fastDummies_1.7.4           lifecycle_1.0.4            </span></span>
<span><span class="co">##   [9] doParallel_1.0.17           globals_0.16.3             </span></span>
<span><span class="co">##  [11] lattice_0.22-6              MASS_7.3-60.2              </span></span>
<span><span class="co">##  [13] magrittr_2.0.3              plotly_4.10.4              </span></span>
<span><span class="co">##  [15] sass_0.4.9                  rmarkdown_2.28             </span></span>
<span><span class="co">##  [17] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span><span class="co">##  [19] httpuv_1.6.15               glmGamPoi_1.16.0           </span></span>
<span><span class="co">##  [21] sctransform_0.4.1           spam_2.10-0                </span></span>
<span><span class="co">##  [23] spatstat.sparse_3.1-0       reticulate_1.39.0          </span></span>
<span><span class="co">##  [25] mapproj_1.2.11              cowplot_1.1.3              </span></span>
<span><span class="co">##  [27] pbapply_1.7-2               RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [29] maps_3.4.2                  zlibbioc_1.50.0            </span></span>
<span><span class="co">##  [31] abind_1.4-5                 GenomicRanges_1.56.1       </span></span>
<span><span class="co">##  [33] Rtsne_0.17                  purrr_1.0.2                </span></span>
<span><span class="co">##  [35] R.utils_2.12.3              GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">##  [37] circlize_0.4.16             IRanges_2.38.1             </span></span>
<span><span class="co">##  [39] S4Vectors_0.42.1            ggrepel_0.9.6              </span></span>
<span><span class="co">##  [41] irlba_2.3.5.1               listenv_0.9.1              </span></span>
<span><span class="co">##  [43] spatstat.utils_3.1-0        goftest_1.2-3              </span></span>
<span><span class="co">##  [45] RSpectra_0.16-2             spatstat.random_3.3-1      </span></span>
<span><span class="co">##  [47] fitdistrplus_1.2-1          parallelly_1.38.0          </span></span>
<span><span class="co">##  [49] DelayedMatrixStats_1.26.0   pkgdown_2.1.0              </span></span>
<span><span class="co">##  [51] DelayedArray_0.30.1         leiden_0.4.3.1             </span></span>
<span><span class="co">##  [53] codetools_0.2-20            tidyselect_1.2.1           </span></span>
<span><span class="co">##  [55] shape_1.4.6.1               farver_2.1.2               </span></span>
<span><span class="co">##  [57] UCSC.utils_1.0.0            matrixStats_1.4.1          </span></span>
<span><span class="co">##  [59] stats4_4.4.1                spatstat.explore_3.3-2     </span></span>
<span><span class="co">##  [61] jsonlite_1.8.8              GetoptLong_1.0.5           </span></span>
<span><span class="co">##  [63] progressr_0.14.0            ggridges_0.5.6             </span></span>
<span><span class="co">##  [65] survival_3.6-4              iterators_1.0.14           </span></span>
<span><span class="co">##  [67] systemfonts_1.1.0           foreach_1.5.2              </span></span>
<span><span class="co">##  [69] tools_4.4.1                 ragg_1.3.2                 </span></span>
<span><span class="co">##  [71] ica_1.0-3                   glue_1.7.0                 </span></span>
<span><span class="co">##  [73] SparseArray_1.4.8           gridExtra_2.3              </span></span>
<span><span class="co">##  [75] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">##  [77] GenomeInfoDb_1.40.1         withr_3.0.1                </span></span>
<span><span class="co">##  [79] BiocManager_1.30.25         fastmap_1.2.0              </span></span>
<span><span class="co">##  [81] fansi_1.0.6                 digest_0.6.37              </span></span>
<span><span class="co">##  [83] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">##  [85] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">##  [87] scattermore_1.2             tensor_1.5                 </span></span>
<span><span class="co">##  [89] dichromat_2.0-0.1           spatstat.data_3.1-2        </span></span>
<span><span class="co">##  [91] R.methodsS3_1.8.2           utf8_1.2.4                 </span></span>
<span><span class="co">##  [93] generics_0.1.3              S4Arrays_1.4.1             </span></span>
<span><span class="co">##  [95] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [97] uwot_0.2.2                  pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [99] gtable_0.3.5                lmtest_0.9-40              </span></span>
<span><span class="co">## [101] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">## [103] dotCall64_1.1-1             clue_0.3-65                </span></span>
<span><span class="co">## [105] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">## [107] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">## [109] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">## [111] rjson_0.2.22                nlme_3.1-164               </span></span>
<span><span class="co">## [113] curl_5.2.2                  cachem_1.1.0               </span></span>
<span><span class="co">## [115] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">## [117] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">## [119] miniUI_0.1.1.1              desc_1.4.3                 </span></span>
<span><span class="co">## [121] pillar_1.9.0                vctrs_0.6.5                </span></span>
<span><span class="co">## [123] promises_1.3.0              xtable_1.8-4               </span></span>
<span><span class="co">## [125] evaluate_0.24.0             magick_2.8.5               </span></span>
<span><span class="co">## [127] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">## [129] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [131] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">## [133] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">## [135] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span><span class="co">## [137] deldir_2.0-4                gridBase_0.4-7             </span></span>
<span><span class="co">## [139] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">## [141] spatstat.geom_3.3-2         RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [143] patchwork_1.2.0             sparseMatrixStats_1.16.0   </span></span>
<span><span class="co">## [145] future_1.34.0               shiny_1.9.1                </span></span>
<span><span class="co">## [147] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [149] ROCR_1.0-11                 gargle_1.5.2               </span></span>
<span><span class="co">## [151] igraph_2.0.3                bslib_0.8.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
