<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/vignettes/Recovery_scRNA.Rmd" class="external-link"><code>vignettes/Recovery_scRNA.Rmd</code></a></small>
      <div class="d-none name"><code>Recovery_scRNA.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial, we will illustrate how to recover spatial ecotypes
(SEs) from single-cell gene expression data, including scRNA-seq and
single-cell-scale spatial transcriptomics datasets generated by
platforms such as MERSCOPE and Xenium.</p>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>The recovery process requires a normalized gene expression matrix and
a vector of cell type annotations.</p>
<details open><summary><strong>Starting from a Seurat object</strong>
</summary><p>A seurat object for the demo data can be accessed from <a href="https://drive.google.com/open?id=11ORWQxxWCNUtceEDwu2IyBpDMa6P-bxS&amp;usp=drive_fs" class="external-link">WU2161_seurat_obj.rds</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"11ORWQxxWCNUtceEDwu2IyBpDMa6P-bxS"</span><span class="op">)</span>, <span class="st">"WU2161_seurat_obj.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>First, load the seurat object into R:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the seurat object</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"WU2161_seurat_obj.rds"</span><span class="op">)</span></span>
<span><span class="va">obj</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 27425 features across 1337 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (27425 features, 2000 variable features)</span></span>
<span><span class="co">##  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cell type annotations</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">obj</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Macrophage" "B"          "CD4T"       "CD8T"       "Other"     </span></span>
<span><span class="co">## [6] "Fibroblast" "Epithelial" "Plasma"</span></span></code></pre>
<p>Next, normalize the gene expression data using <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2" class="external-link">SCTransform</a>
or <a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">NormalizeData</a>.
Here, we are normalizing the data using <code>SCTransform</code>. We
recommend to install the <code>glmGamPoi</code> package for faster
computation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"glmGamPoi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">obj</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Then, extract the normalized gene expression matrix and cell type
annotations from the Seurat object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_version</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">seurat_version</span><span class="op">&lt;</span><span class="fl">5</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">obj</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##            AAACCTGCACATGACT-1 AAACCTGGTGGTCCGT-1 AAACCTGGTTTGTGTG-1</span></span>
<span><span class="co">## AL627309.1                  .                  .                  .</span></span>
<span><span class="co">## AL627309.5                  .                  .                  .</span></span>
<span><span class="co">## AP006222.2                  .                  .                  .</span></span>
<span><span class="co">## LINC01409                   .                  .                  .</span></span>
<span><span class="co">## LINC01128                   .                  .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .                  .</span></span>
<span><span class="co">##            AAACCTGTCCGATATG-1 AAACCTGTCTAACGGT-1</span></span>
<span><span class="co">## AL627309.1                  .                  .</span></span>
<span><span class="co">## AL627309.5                  .                  .</span></span>
<span><span class="co">## AP006222.2                  .                  .</span></span>
<span><span class="co">## LINC01409                   .                  .</span></span>
<span><span class="co">## LINC01128                   .                  .</span></span>
<span><span class="co">## LINC00115                   .                  .</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctann</span> <span class="op">=</span> <span class="va">obj</span><span class="op">$</span><span class="va">CellType</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## AAACCTGCACATGACT-1 AAACCTGGTGGTCCGT-1 AAACCTGGTTTGTGTG-1 AAACCTGTCCGATATG-1 </span></span>
<span><span class="co">##       "Macrophage"                "B"             "CD4T"       "Macrophage" </span></span>
<span><span class="co">## AAACCTGTCTAACGGT-1 AAACGGGCACTTAAGC-1 </span></span>
<span><span class="co">##                "B"             "CD8T"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ctann</span></span>
<span><span class="co">##          B       CD4T       CD8T Epithelial Fibroblast Macrophage      Other </span></span>
<span><span class="co">##         97        138        425        121         44        374         91 </span></span>
<span><span class="co">##     Plasma </span></span>
<span><span class="co">##         47</span></span></code></pre>
</details><details><summary><strong>Starting from Sparse matrix</strong>
</summary><p>Sparse matrix in the .mtx format can be imported using the
<code>ReadMtx</code> function from the <code>Seurat</code> package. The
demo data can be accessed from <a href="https://drive.google.com/open?id=178yvBUNNAAJ8TGtXmvjplVz_Ev7V0WxA&amp;usp=drive_fs" class="external-link">WU2161</a>.
The cell type annotations are available in <a href="https://drive.google.com/open?id=17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ&amp;usp=drive_fs" class="external-link">WU2161_celltype_ann.tsv</a>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load the gene expression matrix</span></span>
<span><span class="va">scdata</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span>mtx <span class="op">=</span> <span class="st">"matrix.mtx"</span>, features <span class="op">=</span> <span class="st">"features.tsv"</span>, </span>
<span>                 cells <span class="op">=</span> <span class="st">"barcodes.tsv"</span>, feature.column <span class="op">=</span> <span class="fl">1</span>, cell.column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Normalize the data</span></span>
<span><span class="va">normdata</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Load cell type annotation</span></span>
<span><span class="va">ctann</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"WU2161_celltype_ann.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                   header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ctann</span> <span class="op">=</span> <span class="va">ctann</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><strong>Starting from tab-delimited files</strong>
</summary><p>Tab-delimited files can be loaded into R using the <code>fread</code>
function from the <code>data.table</code> package. The TSV file for the
expression matrix can be accessed from <a href="https://drive.google.com/open?id=17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei&amp;usp=drive_fs" class="external-link">WU2161_counts.tsv</a>
and the cell type annotations are available in <a href="https://drive.google.com/open?id=17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ&amp;usp=drive_fs" class="external-link">WU2161_celltype_ann.tsv</a>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Download data from google drive</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei"</span><span class="op">)</span>, <span class="st">"WU2161_counts.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17Ax4LMOClMBu6h_WUcwXtFw4HuIU8_AQ"</span><span class="op">)</span>, <span class="st">"WU2161_celltype_ann.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load the gene expression matrix</span></span>
<span><span class="va">scdata</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"WU2161_counts.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">=</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co">## Set the first column as row names</span></span>
<span><span class="va">scdata</span> <span class="op">=</span> <span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">## Drop the first column</span></span>
<span></span>
<span><span class="co">## Normalize the data</span></span>
<span><span class="va">normdata</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            AAACCTGCACATGACT-1 AAACCTGGTGGTCCGT-1 AAACCTGGTTTGTGTG-1</span></span>
<span><span class="co">## AL627309.1                  0                  0                  0</span></span>
<span><span class="co">## AL627309.5                  0                  0                  0</span></span>
<span><span class="co">## AP006222.2                  0                  0                  0</span></span>
<span><span class="co">## AC114498.1                  0                  0                  0</span></span>
<span><span class="co">## AL669831.2                  0                  0                  0</span></span>
<span><span class="co">## LINC01409                   0                  0                  0</span></span>
<span><span class="co">##            AAACCTGTCCGATATG-1 AAACCTGTCTAACGGT-1</span></span>
<span><span class="co">## AL627309.1          0.0000000                  0</span></span>
<span><span class="co">## AL627309.5          0.2951209                  0</span></span>
<span><span class="co">## AP006222.2          0.0000000                  0</span></span>
<span><span class="co">## AC114498.1          0.0000000                  0</span></span>
<span><span class="co">## AL669831.2          0.0000000                  0</span></span>
<span><span class="co">## LINC01409           0.0000000                  0</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load cell type annotation</span></span>
<span><span class="va">ctann</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"WU2161_celltype_ann.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                   header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ctann</span> <span class="op">=</span> <span class="va">ctann</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ctann</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ctann</span></span>
<span><span class="co">##          B       CD4T       CD8T Epithelial Fibroblast Macrophage      Other </span></span>
<span><span class="co">##         97        138        425        121         44        374         91 </span></span>
<span><span class="co">##     Plasma </span></span>
<span><span class="co">##         47</span></span></code></pre>
</details>
</div>
<div class="section level2">
<h2 id="se-recovery">SE recovery<a class="anchor" aria-label="anchor" href="#se-recovery"></a>
</h2>
<p>The <a href="../reference/RecoverSE.html">RecoverSE</a> function will
be used to assign single cells into SEs. Users can either use default
model to recover predefined SEs or use custom model to recover newly
defined SEs.</p>
<p><strong>Note</strong>: Cell type annotation (<code>celltypes</code>
argument) is required for SE recovery using <a href="../reference/RecoverSE.html">RecoverSE</a>.</p>
<details open><summary><strong>Using default models</strong>
</summary><p>The default NMF models were trained on discovery MERSCOPE data,
encompassing five cancer types: melanoma, and four carcinomas. These
models are tailored to nine distinct cell types: B cells, CD4+ T cells,
CD8+ T cells, NK cells, plasma cells, macrophages, dendritic cells,
fibroblasts, and endothelial cells. Each model facilitates the recovery
of SEs from single-cell datasets, allowing for cell-type-specific SE
analysis.</p>
<p>For SE recovery, the cells in the query data should be grouped into
one of “B”, “CD4T”, “CD8T”, “NK”, “Plasma”, “Macrophage”, “DC”,
“Fibroblast”, and “Endothelial”, case sensitive. All the other cell
types will be considered non-SE compartments.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, celltypes <span class="op">=</span> <span class="va">ctann</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sepreds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                   CID   CellType    SE PredScore</span></span>
<span><span class="co">## AAACCTGCACATGACT-1 AAACCTGCACATGACT-1 Macrophage   SE8 0.9958701</span></span>
<span><span class="co">## AAACCTGGTGGTCCGT-1 AAACCTGGTGGTCCGT-1          B NonSE 0.0000000</span></span>
<span><span class="co">## AAACCTGGTTTGTGTG-1 AAACCTGGTTTGTGTG-1       CD4T   SE1 0.9999803</span></span>
<span><span class="co">## AAACCTGTCCGATATG-1 AAACCTGTCCGATATG-1 Macrophage   SE8 0.9920874</span></span>
<span><span class="co">## AAACCTGTCTAACGGT-1 AAACCTGTCTAACGGT-1          B NonSE 0.0000000</span></span>
<span><span class="co">## AAACGGGCACTTAAGC-1 AAACGGGCACTTAAGC-1       CD8T NonSE 0.7479451</span></span></code></pre>
</details><details><summary><strong>Using custom models</strong>
</summary><p>To use custom models, users should first develop recovery models
following the tutorial <a href="TrainRecoveryModels.html">Development of
SE Recovery Models</a>. The resulting models can be used for SE
recovery. An example model is available at <a href="https://drive.google.com/open?id=171WaAe49babYB85Cn1FcoNNE-lzYp1T_&amp;usp=drive_fs" class="external-link">SE_Recovery_W_list.rds</a>.</p>
<p>Download the example models</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"171WaAe49babYB85Cn1FcoNNE-lzYp1T_"</span><span class="op">)</span>, <span class="st">"SE_Recovery_W_list.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Load the custom models</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SE_Recovery_W_list.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">)</span> <span class="co">## named list of W matrices</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CD4T"        "CD8T"        "DC"          "Endothelial" "Fibroblast" </span></span>
<span><span class="co">## [6] "Macrophage"  "NK"          "Plasma"</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">## feature by SE matrix</span></span></code></pre></div>
<pre><code><span><span class="co">##                  SE01      SE02      SE03      SE04      SE05      SE06</span></span>
<span><span class="co">## CD226__pos  0.4145936 0.2555579 0.3323735 0.3241147 0.2263783 0.3279133</span></span>
<span><span class="co">## CDKN1B__pos 0.5729663 0.4552065 0.5120732 0.4627805 0.4453117 0.5156023</span></span>
<span><span class="co">## CXCR4__pos  0.5865790 0.2771225 0.4093821 0.3607076 0.3589891 0.3937211</span></span>
<span><span class="co">## DUSP1__pos  0.4233067 0.3462390 0.2269798 0.1616814 0.3265539 0.2562118</span></span>
<span><span class="co">## KLF2__pos   0.5889618 0.2314586 0.3867504 0.2295206 0.2998234 0.5173246</span></span>
<span><span class="co">## ICAM2__pos  0.4593230 0.2611751 0.4427930 0.3143084 0.3679498 0.4532637</span></span>
<span><span class="co">##                  SE07      SE08      SE09      SE10      SE11</span></span>
<span><span class="co">## CD226__pos  0.2189382 0.3417897 0.2563481 0.2360192 0.2815938</span></span>
<span><span class="co">## CDKN1B__pos 0.4414750 0.4747166 0.4414508 0.3922732 0.3821411</span></span>
<span><span class="co">## CXCR4__pos  0.3751669 0.4456410 0.3023204 0.2629156 0.2729461</span></span>
<span><span class="co">## DUSP1__pos  0.3682236 0.2211759 0.1713413 0.1699067 0.1121763</span></span>
<span><span class="co">## KLF2__pos   0.3570251 0.4327206 0.1732251 0.2086729 0.2118797</span></span>
<span><span class="co">## ICAM2__pos  0.2660441 0.5494585 0.2878338 0.3711269 0.2575091</span></span></code></pre>
<p>Using custom models for SE recovery by specifying the <code>Ws</code>
argument.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, celltypes <span class="op">=</span> <span class="va">ctann</span>, Ws <span class="op">=</span> <span class="va">Ws</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sepreds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                   CID CellType    SE PredScore</span></span>
<span><span class="co">## AAACCTGGTTTGTGTG-1 AAACCTGGTTTGTGTG-1     CD4T  SE01 0.9999803</span></span>
<span><span class="co">## AAAGATGTCATTGCCC-1 AAAGATGTCATTGCCC-1     CD4T  SE09 0.9757931</span></span>
<span><span class="co">## AAAGATGTCCACGTTC-1 AAAGATGTCCACGTTC-1     CD4T  SE10 1.0000000</span></span>
<span><span class="co">## AACCATGAGATGTCGG-1 AACCATGAGATGTCGG-1     CD4T  SE06 0.7376303</span></span>
<span><span class="co">## AACTCCCAGACTACAA-1 AACTCCCAGACTACAA-1     CD4T NonSE 0.4911704</span></span>
<span><span class="co">## AACTGGTTCATCATTC-1 AACTGGTTCATCATTC-1     CD4T  SE02 0.8665178</span></span></code></pre>
</details>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.6.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SpatialEcoTyper_1.0.1 NMF_0.28              Biobase_2.64.0       </span></span>
<span><span class="co">##  [4] BiocGenerics_0.50.0   cluster_2.1.6         rngtools_1.5.2       </span></span>
<span><span class="co">##  [7] registry_0.5-1        RANN_2.6.2            Matrix_1.7-0         </span></span>
<span><span class="co">## [10] googledrive_2.1.1     data.table_1.16.0     Seurat_5.1.0         </span></span>
<span><span class="co">## [13] SeuratObject_5.0.2    sp_2.1-4              ggplot2_3.5.1        </span></span>
<span><span class="co">## [16] dplyr_1.1.4          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">##   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">##   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4             doParallel_1.0.17          </span></span>
<span><span class="co">##   [9] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">##  [11] MASS_7.3-60.2               magrittr_2.0.3             </span></span>
<span><span class="co">##  [13] plotly_4.10.4               sass_0.4.9                 </span></span>
<span><span class="co">##  [15] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span><span class="co">##  [17] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span><span class="co">##  [19] glmGamPoi_1.16.0            sctransform_0.4.1          </span></span>
<span><span class="co">##  [21] spam_2.10-0                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">##  [23] reticulate_1.39.0           cowplot_1.1.3              </span></span>
<span><span class="co">##  [25] pbapply_1.7-2               RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [27] zlibbioc_1.50.0             abind_1.4-5                </span></span>
<span><span class="co">##  [29] GenomicRanges_1.56.1        Rtsne_0.17                 </span></span>
<span><span class="co">##  [31] purrr_1.0.2                 circlize_0.4.16            </span></span>
<span><span class="co">##  [33] GenomeInfoDbData_1.2.12     IRanges_2.38.1             </span></span>
<span><span class="co">##  [35] S4Vectors_0.42.1            ggrepel_0.9.6              </span></span>
<span><span class="co">##  [37] irlba_2.3.5.1               listenv_0.9.1              </span></span>
<span><span class="co">##  [39] spatstat.utils_3.1-0        goftest_1.2-3              </span></span>
<span><span class="co">##  [41] RSpectra_0.16-2             spatstat.random_3.3-1      </span></span>
<span><span class="co">##  [43] fitdistrplus_1.2-1          parallelly_1.38.0          </span></span>
<span><span class="co">##  [45] DelayedMatrixStats_1.26.0   pkgdown_2.1.0              </span></span>
<span><span class="co">##  [47] DelayedArray_0.30.1         leiden_0.4.3.1             </span></span>
<span><span class="co">##  [49] codetools_0.2-20            tidyselect_1.2.1           </span></span>
<span><span class="co">##  [51] shape_1.4.6.1               UCSC.utils_1.0.0           </span></span>
<span><span class="co">##  [53] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">##  [55] spatstat.explore_3.3-2      jsonlite_1.8.8             </span></span>
<span><span class="co">##  [57] GetoptLong_1.0.5            progressr_0.14.0           </span></span>
<span><span class="co">##  [59] ggridges_0.5.6              survival_3.6-4             </span></span>
<span><span class="co">##  [61] iterators_1.0.14            systemfonts_1.1.0          </span></span>
<span><span class="co">##  [63] foreach_1.5.2               tools_4.4.1                </span></span>
<span><span class="co">##  [65] ragg_1.3.2                  ica_1.0-3                  </span></span>
<span><span class="co">##  [67] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">##  [69] SparseArray_1.4.8           gridExtra_2.3              </span></span>
<span><span class="co">##  [71] xfun_0.52                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">##  [73] GenomeInfoDb_1.40.1         withr_3.0.1                </span></span>
<span><span class="co">##  [75] BiocManager_1.30.25         fastmap_1.2.0              </span></span>
<span><span class="co">##  [77] fansi_1.0.6                 digest_0.6.37              </span></span>
<span><span class="co">##  [79] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">##  [81] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">##  [83] scattermore_1.2             tensor_1.5                 </span></span>
<span><span class="co">##  [85] spatstat.data_3.1-2         utf8_1.2.4                 </span></span>
<span><span class="co">##  [87] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">##  [89] S4Arrays_1.4.1              httr_1.4.7                 </span></span>
<span><span class="co">##  [91] htmlwidgets_1.6.4           uwot_0.2.2                 </span></span>
<span><span class="co">##  [93] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">##  [95] ComplexHeatmap_2.20.0       lmtest_0.9-40              </span></span>
<span><span class="co">##  [97] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">##  [99] dotCall64_1.1-1             clue_0.3-65                </span></span>
<span><span class="co">## [101] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">## [103] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">## [105] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">## [107] rjson_0.2.22                nlme_3.1-164               </span></span>
<span><span class="co">## [109] curl_5.2.2                  cachem_1.1.0               </span></span>
<span><span class="co">## [111] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">## [113] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">## [115] miniUI_0.1.1.1              desc_1.4.3                 </span></span>
<span><span class="co">## [117] pillar_1.9.0                grid_4.4.1                 </span></span>
<span><span class="co">## [119] vctrs_0.6.5                 promises_1.3.0             </span></span>
<span><span class="co">## [121] xtable_1.8-4                evaluate_0.24.0            </span></span>
<span><span class="co">## [123] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">## [125] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [127] future.apply_1.11.2         plyr_1.8.9                 </span></span>
<span><span class="co">## [129] fs_1.6.4                    stringi_1.8.4              </span></span>
<span><span class="co">## [131] viridisLite_0.4.2           deldir_2.0-4               </span></span>
<span><span class="co">## [133] gridBase_0.4-7              munsell_0.5.1              </span></span>
<span><span class="co">## [135] lazyeval_0.2.2              spatstat.geom_3.3-2        </span></span>
<span><span class="co">## [137] RcppHNSW_0.6.0              patchwork_1.2.0            </span></span>
<span><span class="co">## [139] sparseMatrixStats_1.16.0    future_1.34.0              </span></span>
<span><span class="co">## [141] shiny_1.9.1                 SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [143] ROCR_1.0-11                 gargle_1.5.2               </span></span>
<span><span class="co">## [145] igraph_2.0.3                bslib_0.8.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
