<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Recovery of Spatial Ecotypes from Spatial Transcriptomics Data • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recovery of Spatial Ecotypes from Spatial Transcriptomics Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Spatial.html">Recovery of Spatial Ecotypes from Spatial Transcriptomics Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Recovery of Spatial Ecotypes from Spatial Transcriptomics Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/vignettes/Recovery_Spatial.Rmd" class="external-link"><code>vignettes/Recovery_Spatial.Rmd</code></a></small>
      <div class="d-none name"><code>Recovery_Spatial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="se-recovery-from-single-cell-spatial-data">SE recovery from single-cell spatial data<a class="anchor" aria-label="anchor" href="#se-recovery-from-single-cell-spatial-data"></a>
</h2>
<p>In this tutorial, we will demonstrate how to recover spatial ecotypes
(SEs) from single-cell spatial transcriptomics data profiled by
platforms such as MERSCOPE, CosMx SMI, or Xenium. We will use a subset
of a melanoma MERSCOPE sample to illustrate the SE recovery process. The
expression data and single-cell metadata can be downloaded from the <a href="https://drive.google.com/open?id=17K5XR7VEvxN9walV4RtaLOR2cBRG-aMh&amp;usp=drive_fs" class="external-link">Google
Drive</a>. In this data, single cells were categorized into nine major
cell types, including B cells, CD4+ T cells, CD8+ T cells, NK cells,
plasma cells, macrophages, dendritic cells (DC), fibroblasts, and
endothelial cells.</p>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p>Download the demo data from Google Drive:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13Rc5Rsu8jbnEYYfUse-xQ7ges51LcI7n"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"12xcZNhpT-xbhcG8kX1QAdTeM9TKeFAUW"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Load the gene expression matrix and meta data into R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell gene expression matrix. Rows are genes, columns are cells.</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, </span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657</span></span>
<span><span class="co">## PDK4                                    0                                1</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     1                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660</span></span>
<span><span class="co">## PDK4                                    1                                0</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     0                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3661</span></span>
<span><span class="co">## PDK4                                    0</span></span>
<span><span class="co">## TNFRSF17                                0</span></span>
<span><span class="co">## ICAM3                                   0</span></span>
<span><span class="co">## FAP                                     0</span></span>
<span><span class="co">## GZMB                                    0</span></span>
<span><span class="co">## TSC2                                    0</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell metadata, with at least three columns, including X, Y, and CellType</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, </span>
<span>                      sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scdata</span> <span class="op">=</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                         X         Y   CellType</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="data-normalization">Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h3>
<p>The gene expression data should be normalized for the SE recovery.
The data can be normalized using <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2" class="external-link">SCTransform</a>
or <a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">NormalizeData</a>.</p>
<p>Here, we are normalizing using SCTransform normalization. We
recommend to install the glmGamPoi package for faster computation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"glmGamPoi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">tmpobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>clip.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_version</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">seurat_version</span><span class="op">&lt;</span><span class="fl">5</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">tmpobj</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="va">tmpobj</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="op">}</span></span></code></pre></div>
<summary>
Using <code>NormalizeData</code> for the normalization
</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="se-recovery">SE recovery<a class="anchor" aria-label="anchor" href="#se-recovery"></a>
</h3>
<p>The <a href="../reference/RecoverSE.html">RecoverSE</a> function will
be used to assign single cells into SEs. Users can either use the
default models to recover predefined SEs or use custom models to recover
newly defined SEs.</p>
<p><strong>Note</strong>: To recover SEs from single-cell data, you must
specify either <code>celltypes</code> or <code>se_results</code> in the
<a href="../reference/RecoverSE.html">RecoverSE</a> function. If neither
is provided, it will assume that the input data represents spot-level
spatial transcriptomics, and SE abundances will be inferred directly
from each spot.</p>
<p>The default NMF models were trained on discovery MERSCOPE data,
encompassing five cancer types: melanoma, and four carcinomas. These
models are tailored to nine distinct cell types: B cells, CD4+ T cells,
CD8+ T cells, NK cells, plasma cells, macrophages, dendritic cells,
fibroblasts, and endothelial cells. Each model facilitates the recovery
of SEs from single-cell datasets, allowing for cell-type-specific SE
analysis. Thus, for SE recovery using default models, the cells in the
query data should be grouped into one of “B”, “CD4T”, “CD8T”, “NK”,
“Plasma”, “Macrophage”, “DC”, “Fibroblast”, and “Endothelial”, case
sensitive. All the other cell types will be considered non-SE
compartments.</p>
<details open><summary><strong>Using default models (1)</strong>
</summary><p>Before SE recovery, we recommend to use a unified embedding of
spatial microregions by performing de novo Spatial EcoTyper analysis,
which integrate gene expression and spatial information. This embedding
could enhance the accuracy and refinement of SE recovery results. The
detailed tutorial for Spatial EcoTyper analysis is available at <a href="SingleSample.html">Discovery of Spatial Ecotypes from a
Single-cell Spatial Dataset</a>.</p>
<p>For demonstration purposes, we used <code>SpatialEcoTyper</code> to
group cells into spatial clusters with a resolution of 10. In practice,
we recommend experimenting with multiple resolutions to ensure robust
and reliable results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## make sure the cells are grouped into one of "B", "CD4T", "CD8T", "NK", "Plasma", "Macrophage", "DC", "Fibroblast", and "Endothelial".</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Fibroblast"  "Endothelial" "DC"          "Macrophage"  "CD8T"       </span></span>
<span><span class="co">## [6] "CD4T"        "Plasma"      "NK"          "B"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Spatial EcoTyper analysis: it would take ~2 min</span></span>
<span><span class="va">emb</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a></span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>, resolution <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">emb</span><span class="op">$</span><span class="va">obj</span> <span class="co">## A seurat object including the spatial embedding</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 2315 features across 2315 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (2315 features, 2315 variable features)</span></span>
<span><span class="co">##  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">emb</span><span class="op">$</span><span class="va">metadata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span>, <span class="st">"SE"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>  <span class="co">## Single-cell meta data with SE annotations</span></span></code></pre></div>
<pre><code><span><span class="co">##                                         X         Y   CellType   SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast SE25</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast SE29</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast SE29</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast SE29</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast SE48</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast SE25</span></span></code></pre>
<p>Then specify the <code>se_results</code> in the <a href="../reference/RecoverSE.html">RecoverSE</a> function for SE
recovery.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, se_results <span class="op">=</span> <span class="va">emb</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sepreds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657 </span></span>
<span><span class="co">##                            "SE1"                            "SE4" </span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660 </span></span>
<span><span class="co">##                            "SE4"                            "SE4" </span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 HumanMelanomaPatient1__cell_3663 </span></span>
<span><span class="co">##                            "SE3"                            "SE1"</span></span></code></pre>
</details><details><summary><strong>Using default models (2)</strong>
</summary><p>You can also recover the SEs without using spatial embedding, which
could be less accurate due to the lack of spatial information. The cell
type annotations are required in this case. Cells should be grouped into
one of “B”, “CD4T”, “CD8T”, “NK”, “Plasma”, “Macrophage”, “DC”,
“Fibroblast”, and “Endothelial”, case sensitive. All the other cell
types will be considered non-SE compartments.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## make sure the cells are grouped into one of "B", "CD4T", "CD8T", "NK", "Plasma", "Macrophage", "DC", "Fibroblast", and "Endothelial".</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, celltypes <span class="op">=</span> <span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sepreds</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><strong>Using custom models</strong>
</summary><p>To use custom models, users should first develop recovery models
following the tutorial <a href="TrainRecoveryModels.html">Development of
SE Recovery Models</a>. The resulting models can be used for SE
recovery. An example model is available at <a href="https://drive.google.com/open?id=171WaAe49babYB85Cn1FcoNNE-lzYp1T_&amp;usp=drive_fs" class="external-link">SE_Recovery_W_list.rds</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download SE recovery model</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"171WaAe49babYB85Cn1FcoNNE-lzYp1T_"</span><span class="op">)</span>, <span class="st">"SE_Recovery_W_list.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Using custom models by specifying the Ws argument:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SE_Recovery_W_list.rds"</span><span class="op">)</span></span>
<span><span class="co">## make sure the cell type names are consistent.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, celltypes <span class="op">=</span> <span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span>, Ws <span class="op">=</span> <span class="va">Ws</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## If Spatial EcoTyper result is available, we recommend:</span></span>
<span><span class="va">sepreds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, se_results <span class="op">=</span> <span class="va">emb</span>, Ws <span class="op">=</span> <span class="va">Ws</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="visualization-of-ses-in-the-tissue">Visualization of SEs in the tissue<a class="anchor" aria-label="anchor" href="#visualization-of-ses-in-the-tissue"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add the recovery result into the meta data</span></span>
<span><span class="va">scmeta</span><span class="op">$</span><span class="va">RecoveredSE</span> <span class="op">&lt;-</span> <span class="va">sepreds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Visualize the SE recovery results</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta</span>, by <span class="op">=</span> <span class="st">"RecoveredSE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Recovery_Spatial_files/figure-html/visualscst-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="recovery-of-spatial-ecotypes-from-visium-spatial-gene-expression-data">Recovery of spatial ecotypes from Visium Spatial Gene Expression
data<a class="anchor" aria-label="anchor" href="#recovery-of-spatial-ecotypes-from-visium-spatial-gene-expression-data"></a>
</h2>
<p>To recover SEs from Visium spatial transcriptomics data, we first
infer SE abundances from each spot. Each spot is then assigned to the SE
with the highest inferred abundance, enabling the spatial mapping of SEs
across the tissue. In this tutorial, we will use a breast cancer sample
to demonstrate the SE recovery from a Visium data. The expression data
can be accessed from: <a href="https://drive.google.com/open?id=15D9LgvZmCZUfsL62cD67JMf8Jcq_UyuB&amp;usp=drive_fs" class="external-link">V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5</a>,
which was downloaded from <a href="https://www.10xgenomics.com/datasets" class="external-link">10x Genomics
datasets</a>.</p>
<div class="section level3">
<h3 id="loading-data-1">Loading data<a class="anchor" aria-label="anchor" href="#loading-data-1"></a>
</h3>
<p>First, download the data from Google Drive</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"15D9LgvZmCZUfsL62cD67JMf8Jcq_UyuB"</span><span class="op">)</span>, <span class="st">"V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"15NTZc1HrW_gLS_pmi1ckYLw30kNarf4w"</span><span class="op">)</span>, <span class="st">"V1_Breast_Cancer_Block_A_Section_1_tissue_positions_list.csv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## This download should be done within 1 min.</span></span></code></pre></div>
<p>Load the expression data into R using the <code>Read10X_h5</code>
function from the Seurat package.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"hdf5r"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"hdf5r"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://hhoeflin.github.io/hdf5r/" class="external-link">"hdf5r"</a></span><span class="op">)</span></span>
<span><span class="co"># Load Visium gene expression matrix. Rows are genes, columns are spots.</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="co"># normalize the expression data</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"V1_Breast_Cancer_Block_A_Section_1_tissue_positions_list.csv"</span>, </span>
<span>                 header <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tissue"</span>, <span class="st">"row"</span>, <span class="st">"col"</span>, <span class="st">"imagerow"</span>, <span class="st">"imagecol"</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    tissue row col imagerow imagecol</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1      1  50 102    15632    17782</span></span>
<span><span class="co">## AAACACCAATAACTGC-1      1  59  19    17734     6447</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1      1  14  94     7079    16716</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1      1  47  13    14882     5637</span></span>
<span><span class="co">## AAACAGTGTTCCTGGG-1      1  73  43    21069     9712</span></span>
<span><span class="co">## AAACATTTCCCGGATT-1      1  61  97    18242    17091</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="se-recovery-1">SE recovery<a class="anchor" aria-label="anchor" href="#se-recovery-1"></a>
</h3>
<p>If neither <code>celltypes</code> nor <code>se_results</code> are
specified, the <a href="../reference/RecoverSE.html">RecoverSE</a>
function will assume the input data is a gene-by-spot gene expression
matrix and infer SE abundances across spots. Users have the option to
either use the default models to recover predefined SEs or apply custom
models to recover newly defined SEs.</p>
<details open><summary><strong>Using default models</strong>
</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">preds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">preds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                             SE1          SE2          SE3          SE4</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1 2.127701e-04 6.877800e-05 3.979458e-01 8.534547e-02</span></span>
<span><span class="co">## AAACACCAATAACTGC-1 1.321616e-01 6.347672e-02 4.580604e-02 9.893653e-02</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1 1.658118e-01 2.438595e-15 1.257332e-01 1.021615e-01</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1 6.495088e-02 1.975689e-16 2.208619e-01 6.972421e-10</span></span>
<span><span class="co">## AAACAGTGTTCCTGGG-1 9.104830e-02 2.073793e-01 2.579438e-09 6.489594e-02</span></span>
<span><span class="co">## AAACATTTCCCGGATT-1 6.684999e-13 2.987802e-01 2.908525e-08 7.202254e-07</span></span>
<span><span class="co">##                             SE5          SE6          SE7          SE8</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1 1.162860e-01 1.723712e-16 3.026670e-01 2.078080e-02</span></span>
<span><span class="co">## AAACACCAATAACTGC-1 4.542600e-07 4.511416e-01 3.746409e-16 1.333849e-01</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1 7.829883e-03 3.928577e-05 3.462729e-01 2.635185e-04</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1 8.604433e-03 1.976922e-16 7.055828e-01 1.739479e-09</span></span>
<span><span class="co">## AAACAGTGTTCCTGGG-1 1.529680e-06 2.947065e-01 2.356171e-15 1.632831e-01</span></span>
<span><span class="co">## AAACATTTCCCGGATT-1 1.993179e-11 1.996192e-01 3.022784e-16 2.576151e-01</span></span>
<span><span class="co">##                             SE9        nonSE</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1 1.755819e-02 5.913523e-02</span></span>
<span><span class="co">## AAACACCAATAACTGC-1 4.131729e-03 7.096041e-02</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1 1.684507e-02 2.350429e-01</span></span>
<span><span class="co">## AAACAGGGTCTATATT-1 1.975689e-16 2.054490e-11</span></span>
<span><span class="co">## AAACAGTGTTCCTGGG-1 2.351639e-06 1.786829e-01</span></span>
<span><span class="co">## AAACATTTCCCGGATT-1 7.419167e-02 1.697931e-01</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## This step would take ~3 minutes to complete</span></span></code></pre></div>
</details><details><summary><strong>Using custom models</strong>
</summary><p>To develop SE recovery models, users should follow the tutorial <a href="TrainRecoveryModels.html">Developing SE Recovery Models</a>. The
resulting models can be used for SE recovery. An example model is
available at <a href="https://drive.google.com/open?id=171WaAe49babYB85Cn1FcoNNE-lzYp1T_&amp;usp=drive_fs" class="external-link">SE_Recovery_W_list.rds</a>.
By specifying the <code>Ws</code> parameter (a list of
cell-type-specific W matrices) in the <a href="../reference/RecoverSE.html">RecoverSE</a> function, the custom
models will be used for recovering SEs.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download SE recovery model</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"171WaAe49babYB85Cn1FcoNNE-lzYp1T_"</span><span class="op">)</span>, <span class="st">"SE_Recovery_W_list.rds"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"SE_Recovery_W_list.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Using custom models by specifying the Ws argument</span></span>
<span><span class="va">preds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RecoverSE.html">RecoverSE</a></span><span class="op">(</span><span class="va">normdata</span>, Ws <span class="op">=</span> <span class="va">Ws</span><span class="op">)</span></span>
<span><span class="co">## This step would take ~3 minutes to complete</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="visualization-of-ses-in-the-tissue-1">Visualization of SEs in the tissue<a class="anchor" aria-label="anchor" href="#visualization-of-ses-in-the-tissue-1"></a>
</h3>
<p>Each spot can be assigned to the SE with the highest inferred
abundance, and the spatial mapping of SEs can be visualized using the <a href="../reference/SpatialView.html">SpatialView</a> function.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span><span class="op">$</span><span class="va">RecoveredSE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">preds</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">preds</span>, <span class="fl">1</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">Y</span> <span class="op">=</span> <span class="op">-</span><span class="va">meta</span><span class="op">$</span><span class="va">row</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">meta</span>, by <span class="op">=</span> <span class="st">"RecoveredSE"</span>, X <span class="op">=</span> <span class="st">"col"</span>, Y <span class="op">=</span> <span class="st">"Y"</span>, pt.size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Recovery_Spatial_files/figure-html/visualvisium-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] hdf5r_1.3.11          SpatialEcoTyper_1.0.0 NMF_0.28             </span></span>
<span><span class="co">##  [4] Biobase_2.64.0        BiocGenerics_0.50.0   cluster_2.1.6        </span></span>
<span><span class="co">##  [7] rngtools_1.5.2        registry_0.5-1        RANN_2.6.2           </span></span>
<span><span class="co">## [10] Matrix_1.7-0          googledrive_2.1.1     data.table_1.16.0    </span></span>
<span><span class="co">## [13] Seurat_5.1.0          SeuratObject_5.0.2    sp_2.1-4             </span></span>
<span><span class="co">## [16] ggplot2_3.5.1         dplyr_1.1.4          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">##   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">##   [5] R.oo_1.26.0                 polyclip_1.10-7            </span></span>
<span><span class="co">##   [7] fastDummies_1.7.4           lifecycle_1.0.4            </span></span>
<span><span class="co">##   [9] doParallel_1.0.17           pals_1.9                   </span></span>
<span><span class="co">##  [11] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">##  [13] MASS_7.3-60.2               magrittr_2.0.3             </span></span>
<span><span class="co">##  [15] plotly_4.10.4               sass_0.4.9                 </span></span>
<span><span class="co">##  [17] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span><span class="co">##  [19] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span><span class="co">##  [21] glmGamPoi_1.16.0            sctransform_0.4.1          </span></span>
<span><span class="co">##  [23] spam_2.10-0                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">##  [25] reticulate_1.39.0           mapproj_1.2.11             </span></span>
<span><span class="co">##  [27] cowplot_1.1.3               pbapply_1.7-2              </span></span>
<span><span class="co">##  [29] RColorBrewer_1.1-3          maps_3.4.2                 </span></span>
<span><span class="co">##  [31] zlibbioc_1.50.0             abind_1.4-5                </span></span>
<span><span class="co">##  [33] GenomicRanges_1.56.1        Rtsne_0.17                 </span></span>
<span><span class="co">##  [35] purrr_1.0.2                 R.utils_2.12.3             </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.12     circlize_0.4.16            </span></span>
<span><span class="co">##  [39] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">##  [41] ggrepel_0.9.6               irlba_2.3.5.1              </span></span>
<span><span class="co">##  [43] listenv_0.9.1               spatstat.utils_3.1-0       </span></span>
<span><span class="co">##  [45] goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span><span class="co">##  [47] spatstat.random_3.3-1       fitdistrplus_1.2-1         </span></span>
<span><span class="co">##  [49] parallelly_1.38.0           DelayedMatrixStats_1.26.0  </span></span>
<span><span class="co">##  [51] pkgdown_2.1.0               DelayedArray_0.30.1        </span></span>
<span><span class="co">##  [53] leiden_0.4.3.1              codetools_0.2-20           </span></span>
<span><span class="co">##  [55] tidyselect_1.2.1            shape_1.4.6.1              </span></span>
<span><span class="co">##  [57] farver_2.1.2                UCSC.utils_1.0.0           </span></span>
<span><span class="co">##  [59] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">##  [61] spatstat.explore_3.3-2      jsonlite_1.8.8             </span></span>
<span><span class="co">##  [63] GetoptLong_1.0.5            progressr_0.14.0           </span></span>
<span><span class="co">##  [65] ggridges_0.5.6              survival_3.6-4             </span></span>
<span><span class="co">##  [67] iterators_1.0.14            systemfonts_1.1.0          </span></span>
<span><span class="co">##  [69] foreach_1.5.2               tools_4.4.1                </span></span>
<span><span class="co">##  [71] ragg_1.3.2                  ica_1.0-3                  </span></span>
<span><span class="co">##  [73] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">##  [75] SparseArray_1.4.8           gridExtra_2.3              </span></span>
<span><span class="co">##  [77] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">##  [79] GenomeInfoDb_1.40.1         withr_3.0.1                </span></span>
<span><span class="co">##  [81] BiocManager_1.30.25         fastmap_1.2.0              </span></span>
<span><span class="co">##  [83] fansi_1.0.6                 digest_0.6.37              </span></span>
<span><span class="co">##  [85] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">##  [87] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">##  [89] scattermore_1.2             tensor_1.5                 </span></span>
<span><span class="co">##  [91] dichromat_2.0-0.1           spatstat.data_3.1-2        </span></span>
<span><span class="co">##  [93] R.methodsS3_1.8.2           utf8_1.2.4                 </span></span>
<span><span class="co">##  [95] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">##  [97] S4Arrays_1.4.1              httr_1.4.7                 </span></span>
<span><span class="co">##  [99] htmlwidgets_1.6.4           uwot_0.2.2                 </span></span>
<span><span class="co">## [101] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">## [103] ComplexHeatmap_2.20.0       lmtest_0.9-40              </span></span>
<span><span class="co">## [105] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">## [107] dotCall64_1.1-1             clue_0.3-65                </span></span>
<span><span class="co">## [109] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">## [111] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">## [113] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">## [115] rjson_0.2.22                nlme_3.1-164               </span></span>
<span><span class="co">## [117] curl_5.2.2                  cachem_1.1.0               </span></span>
<span><span class="co">## [119] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">## [121] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">## [123] miniUI_0.1.1.1              desc_1.4.3                 </span></span>
<span><span class="co">## [125] pillar_1.9.0                grid_4.4.1                 </span></span>
<span><span class="co">## [127] vctrs_0.6.5                 promises_1.3.0             </span></span>
<span><span class="co">## [129] xtable_1.8-4                evaluate_0.24.0            </span></span>
<span><span class="co">## [131] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">## [133] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [135] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">## [137] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">## [139] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span><span class="co">## [141] deldir_2.0-4                gridBase_0.4-7             </span></span>
<span><span class="co">## [143] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">## [145] spatstat.geom_3.3-2         RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [147] patchwork_1.2.0             bit64_4.5.2                </span></span>
<span><span class="co">## [149] sparseMatrixStats_1.16.0    future_1.34.0              </span></span>
<span><span class="co">## [151] shiny_1.9.1                 highr_0.11                 </span></span>
<span><span class="co">## [153] SummarizedExperiment_1.34.0 ROCR_1.0-11                </span></span>
<span><span class="co">## [155] gargle_1.5.2                igraph_2.0.3               </span></span>
<span><span class="co">## [157] bslib_0.8.0                 bit_4.5.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
