<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Discovery of Spatial Ecotypes from a Single Sample • SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Discovery of Spatial Ecotypes from a Single Sample">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Discovery of Spatial Ecotypes from a Single Sample</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/vignettes/SingleSample.Rmd" class="external-link"><code>vignettes/SingleSample.Rmd</code></a></small>
      <div class="d-none name"><code>SingleSample.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this tutorial, we will illustrate how to perform de novo discovery
of spatial ecotypes from single-cell spatial transcriptomics dataset
using <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>.</p>
<p>We will be analyzing single-cell spatial transcriptomics data from a
melanoma sample (raw data available in <a href="https://vizgen.com/data-release-program/" class="external-link">Vizgen’s MERSCOPE FFPE
Human Immuno-oncology</a>). To quickly demonstrate how <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a> works,
we’ve selected a subset of the sample, which can be downloaded from <a href="https://drive.google.com/drive/folders/17K5XR7VEvxN9walV4RtaLOR2cBRG-aMh?usp=drive_link" class="external-link"><code>here</code></a>.</p>
<p>This demo data comprises the spatial expression of 500 genes across
27,907 cells, which are categorized into ten distinct cell types: B
cells, CD4+ T cells, CD8+ T cells, NK cells, plasma cells, macrophages,
dendritic cells (DC), fibroblasts, endothelial cells, and melanoma
cells. Melanoma cells are excluded from this demonstration to reduce
processing time.</p>
<p>All cells are grouped into four spatial regions: tumor, inner margin,
outer margin, and stroma. The tumor and stroma regions are defined based
on the density of cancer cells, as described in the <a href="https://www.nature.com/articles/s41587-023-01697-9" class="external-link">CytoSPACE
paper</a>. The inner and outer margins are defined as regions extending
250 μm inside and outside the tumor boundaries, respectively.
Furthermore, we quantified each cell’s distance to the tumor–stroma
interface by calculating the shortest Euclidean distance to the nearest
tumor region (for stromal cells) or stromal region (for tumor cells). A
positive distance indicates cells located within the tumor region, while
a negative distance indicates cells located within the stroma.</p>
<p><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>
analysis requires two input data:</p>
<ul>
<li>
<strong>gene expression matrix</strong>: rows represent gene names
and columns represent cell IDs</li>
<li>
<strong>meta data</strong>: a data frame with at least three
columns, including “X” (X-coordinate), “Y” (Y-coordinate), and
“CellType” (the cell type annotation). The row names of the meta data
should match the column names (cell IDs) in the expression matrix.</li>
</ul>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">R.utils</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Download an example data for the analysis</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13Rc5Rsu8jbnEYYfUse-xQ7ges51LcI7n"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"12xcZNhpT-xbhcG8kX1QAdTeM9TKeFAUW"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load the gene expression matrix</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, </span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Normalize the gene expression data</span></span>
<span><span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load the spatial coordinates and cell type annotation of single cells</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, </span>
<span>                      sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">scmeta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## Discover spatial ecotypes using the SpatialEcoTyper</span></span>
<span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a></span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>, </span>
<span>                              outprefix <span class="op">=</span> <span class="st">"Melanoma1_subset"</span>,</span>
<span>                              radius <span class="op">=</span> <span class="fl">50</span>, ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<details open><summary><strong>Text files as input</strong>
</summary><p>Download the data from Google Drive</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13Rc5Rsu8jbnEYYfUse-xQ7ges51LcI7n"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"12xcZNhpT-xbhcG8kX1QAdTeM9TKeFAUW"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Large text files can be loaded into R using the <code>fread</code>
function from the <code>data.table</code> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell gene expression matrix. Rows represent gene names and columns represent cell IDs</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, </span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co"># Setting the first column as row names</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># Dropping first column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657</span></span>
<span><span class="co">## PDK4                                    0                                1</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     1                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660</span></span>
<span><span class="co">## PDK4                                    1                                0</span></span>
<span><span class="co">## TNFRSF17                                0                                0</span></span>
<span><span class="co">## ICAM3                                   0                                0</span></span>
<span><span class="co">## FAP                                     0                                0</span></span>
<span><span class="co">## GZMB                                    0                                0</span></span>
<span><span class="co">## TSC2                                    0                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3661</span></span>
<span><span class="co">## PDK4                                    0</span></span>
<span><span class="co">## TNFRSF17                                0</span></span>
<span><span class="co">## ICAM3                                   0</span></span>
<span><span class="co">## FAP                                     0</span></span>
<span><span class="co">## GZMB                                    0</span></span>
<span><span class="co">## TSC2                                    0</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell metadata</span></span>
<span><span class="co"># Three columns are required, including "X", "Y", and "CellType"</span></span>
<span><span class="co"># The row names should match the cell ids in the expression matrix</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>, </span>
<span>                      sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">scmeta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span> <span class="co"># match the cell ids in scdata and scmeta</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"CellType"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                         X         Y   CellType</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast</span></span></code></pre>
</details><details><summary><strong>Sparse matrix as input</strong>
</summary><p><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>
supports sparse matrix as input. Mtx files can be loaded into R using
the <a href="https://satijalab.org/seurat/reference/readmtx" class="external-link"><code>ReadMtx</code></a>
function from the Seurat package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13M3xhRxp0xK9gf5F4DE9idSBFqVQIXDT"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.mtx.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"136feRaFjMtNvduLTm5xqa3WhyyoG4Xzo"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_cells.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13QprWzJhzzUy_w3XSrjlt9pjf2n-G7HV"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_genes.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span>mtx <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_counts.mtx.gz"</span>, </span>
<span>                  cells <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_cells.tsv.gz"</span>, </span>
<span>                  features <span class="op">=</span> <span class="st">"HumanMelanomaPatient1_subset_genes.tsv.gz"</span>,</span>
<span>                  feature.column <span class="op">=</span> <span class="fl">1</span>, cell.column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="data-normalization">Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h2>
<p>The gene expression data should be normalized before <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a> analysis.
The data can be normalized using <a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">NormalizeData</a>
or <a href="https://satijalab.org/seurat/articles/seurat5_spatial_vignette_2" class="external-link">SCTransform</a>.</p>
<p>Here, we are normalizing using SCTransform normalization. We
recommend to install the glmGamPoi package for faster computation.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"glmGamPoi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"glmGamPoi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">tmpobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>clip.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_version</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">seurat_version</span><span class="op">&lt;</span><span class="fl">5</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">tmpobj</span>, <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="va">normdata</span> <span class="op">&lt;-</span> <span class="va">tmpobj</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="op">}</span></span></code></pre></div>
<details><summary><strong>Using <code>NormalizeData</code> for the normalization:</strong>
</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="preview-of-the-sample">Preview of the sample<a class="anchor" aria-label="anchor" href="#preview-of-the-sample"></a>
</h2>
<p>The <a href="../reference/SpatialView.html">SpatialView</a> function
can be used to visualize single cells within the tissue. You can color
the cells by cell type or predefined spatial regions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the cell type annotations in the tissue</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta</span>, by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/viewct-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the regions in the tissue</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta</span>, by <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/viewregion-1.png" width="700"></p>
<p>The <a href="../reference/SpatialView.html">SpatialView</a> function
can also be used to visualize continuous characteristics, such as the
minimum distance of each single cell to tumor/stroma margin. Here,
positive distances indicate cells located within the tumor region, while
negative distances denote cells within the stroma.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize the distance to tumor margin</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta</span>, by <span class="op">=</span> <span class="st">"Dist2Interface"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#5e3c99"</span>, high <span class="op">=</span> <span class="st">"#e66101"</span>, </span>
<span>                         mid <span class="op">=</span> <span class="st">"#d9d9d9"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Distance to\ntumor margin"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/viewdist-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="se-discovery-using-spatialecotyper">SE discovery using SpatialEcoTyper<a class="anchor" aria-label="anchor" href="#se-discovery-using-spatialecotyper"></a>
</h2>
<p><img src="images/SpatialEcoTyper_Workflow.png" width="100%"></p>
<p>The <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>
function is designed to identify spatial ecotypes (SEs) from single-cell
spatial transcriptomics data. The workflow begins by defining spatial
neighborhoods (SNs) on a regular grid and constructing cell
type–specific gene expression profiles (GEPs) for each SN. For each cell
type, a similarity network is then generated based on these GEPs, where
nodes represent SNs and edges denote their expression similarity. To
integrate these networks across cell types, SpatialEcoTyper applies the
Similarity Network Fusion (SNF) method, originally developed for
multi-omics data integration (<a href="https://www.nature.com/articles/nmeth.2810" class="external-link">Wang et al.,
2014</a>). This yields a fused similarity network that captures shared
spatial transcriptomic variation across cell types, enabling the
identification of SEs through clustering.</p>
<details open><summary><strong>Key arguments for SpatialEcoTyper</strong>
</summary><ul>
<li>
<code>normdata</code> A matrix representing normalized gene
expression data, where rows correspond to genes and columns correspond
to cells.</li>
<li>
<code>metadata</code> A data frame containing metadata associated
with each cell. Must include spatial coordinates (e.g., X and Y) as well
as cell type annotations. The row names must match the column names of
the <code>normdata</code>.</li>
<li>
<code>outprefix</code> Character string specifying the prefix for
output file names.</li>
<li>
<code>radius</code> Numeric specifying the radius (default: 50 µm)
for defining spatial neighborhood.</li>
<li>
<code>grid.size</code> Numeric specifying the grid size for spatial
discretization of coordinates.</li>
<li>
<code>resolution</code> Numeric specifying the resolution for
Louvain clustering (default: 0.5).</li>
<li>
<code>nfeatures</code> Integer specifying the number of top variable
features (default: 500) used for the analysis.</li>
<li>
<code>min.cts.per.region</code> Integer specifying the minimum
number of cell types (default: 2) required for a spatial
neighborhood.</li>
<li>
<code>iterations</code> Integer specifying the number of iterations
(default: 10) for SNF analysis.</li>
<li>
<code>ncores</code> Integer specifying the number of CPU cores to
use for parallel processing.</li>
</ul>
<p>You can type <code><a href="../reference/SpatialEcoTyper.html">?SpatialEcoTyper</a></code> to visualize the full
manual.</p>
</details><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## This step takes ~2 minutes to complete on macOS with an Apple M1 Pro chip and 16 GB memory.</span></span>
<span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a></span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>,</span>
<span>                              outprefix <span class="op">=</span> <span class="st">"Melanoma1_subset"</span>,</span>
<span>                              radius <span class="op">=</span> <span class="fl">50</span>, ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>Optimizing memory usage</strong>
</summary><p>For large single-cell spatial transcriptomics (ST) datasets with more
than 100,000 cells, the analysis can be both time- and memory-intensive.
To accelerate computation, you can increase the number of cores used
(<code>ncores</code>), but this will also increase memory
consumption.</p>
<p>If computational memory is limited, several strategies can help
reduce usage. One option is to increase the <code>grid.size</code>,
which decreases the number of spatial neighborhoods and can
substantially decrease memory usage. Another option is to reduce the
<code>minibatch</code> size and the number of cores
(<code>ncores</code>). However, the minimum memory requirement may still
remain high for very large datasets.</p>
</details><details><summary><strong>Interested in multicellular communities associated with specific
cell types?</strong>
</summary><p>If you’re interested in investigating multicellular communities
associated with specific cell types, you can use the
<code>filter.region.by.celltypes</code> argument to limit the analysis
to spatial neighborhoods that contain at least one cell of the specified
type. For example, to identify SEs associated with CD8 T cells, you can
use the following command:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a></span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>,</span>
<span>                              outprefix <span class="op">=</span> <span class="st">"Melanoma1_subset"</span>,</span>
<span>                              radius <span class="op">=</span> <span class="fl">50</span>, ncores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                              filter.region.by.celltypes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8T"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><strong>Interested in regions with multiple cell types?</strong>
</summary><p>If you’re interested in regions composed of multiple cell types, you
can use the <code>min.cts.per.region</code> argument to restrict the
analysis to spatial neighborhoods containing at least the specified
minimum number of distinct cell types.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a></span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>,</span>
<span>                              outprefix <span class="op">=</span> <span class="st">"Melanoma1_subset"</span>,</span>
<span>                              radius <span class="op">=</span> <span class="fl">50</span>, ncores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                              min.cts.per.region <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="spatialecotyper-result">SpatialEcoTyper result<a class="anchor" aria-label="anchor" href="#spatialecotyper-result"></a>
</h2>
<p>When the <code>outprefix</code> is specified, the <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a> result will
be saved as a RDS file named
<code>outprefix_SpatialEcoTyper_results.rds</code>. The result can be
loaded into R using <code>readRDS</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## drive_download(as_id("13T8IIbgk26sokFI8F3VqI4mvz3QIp3fT"), "Melanoma1_subset_SpatialEcoTyper_results.rds", overwrite = TRUE)</span></span>
<span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Melanoma1_subset_SpatialEcoTyper_results.rds"</span><span class="op">)</span></span></code></pre></div>
<p>The <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>
result is a list that includes two key components:</p>
<ul>
<li>
<strong>a Seurat object</strong> constructed from the fused network
embedding of spatial neighborhoods, enabling the identification of
spatial ecotypes through clustering.</li>
<li>
<strong>a data frame of the single-cell metadata</strong>, which
includes the original single-cell metadata with spatial ecotype cluster
annotations added.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the Seurat object and updated single-cell metadata</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">$</span><span class="va">obj</span> <span class="co"># A Seurat object</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">$</span><span class="va">metadata</span> <span class="co"># Single-cell meta data, with SE annotation added</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                         X         Y   CellType CellTypeName</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast  Fibroblasts</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast  Fibroblasts</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast  Fibroblasts</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast  Fibroblasts</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast  Fibroblasts</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast  Fibroblasts</span></span>
<span><span class="co">##                                  Region Dist2Interface  SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655 Stroma      -883.1752 SE3</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657 Stroma      -894.8463 SE6</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658 Stroma      -904.1115 SE6</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660 Stroma      -907.8909 SE6</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661 Stroma      -874.2712 SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663 Stroma      -903.6559 SE3</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span> <span class="co">## The number of cells in each SE</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  SE0  SE1  SE2  SE3  SE4  SE5  SE6  SE7 </span></span>
<span><span class="co">## 4067 2480 6907 4297 3952 1232 1801 2244</span></span></code></pre>
<p><strong>Note</strong>: Spatial EcoTyper employs stringent quality
control steps to exclude spatial neighborhoods with suboptimal gene
expression profiles, such as those with insufficiently detected genes
(<code>min.feature</code> parameter), insufficient cells
(<code>min.cells</code> parameter), and underrepresented cell types
(<code>min.cts.per.region</code> parameter). These low confidence
regions are marked as NA by the pipeline. While regions designated as NA
at one resolution (<code>radius</code> and <code>grid.size</code>
parameter) may be included at a lower resolution if the larger region
passes QC, we observed a negligible impact of different resolutions on
overall performance. By default, NA regions are removed from the output
to avoid confusion. To retain all cells in the single-cell meta data,
you can set <code>dropcell = FALSE</code> in the <a href="../reference/SpatialEcoTyper.html">SpatialEcoTyper</a>.</p>
</div>
<div class="section level2">
<h2 id="embedding-of-spatial-architecture">Embedding of spatial architecture<a class="anchor" aria-label="anchor" href="#embedding-of-spatial-architecture"></a>
</h2>
<p>The embedding of spatial neighborhoods can be visualized using
standard <a href="https://satijalab.org/seurat/articles/visualization_vignette" class="external-link">Seurat
functions</a> such as <code>DimPlot</code> and <code>FeaturePlot</code>.
These visualizations help to explore the spatial organization and
heterogeneity within the tissue.</p>
<p><strong>Note</strong>: The embedding and clustering results differ
slightly between Seurat v4 and v5. However, the overall clustering
patterns remain largely consistent, with an Adjusted Rand Index (ARI) of
0.7 for the demonstration dataset. This consistency ensures that,
despite minor variations, the key biological insights are preserved
across versions.</p>
<div class="section level3">
<h3 id="visualizing-tumorstroma-regions-in-the-embedding">Visualizing tumor/stroma regions in the embedding<a class="anchor" aria-label="anchor" href="#visualizing-tumorstroma-regions-in-the-embedding"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/umap_region-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-the-distance-of-sns-to-tumorstroma-interface">Visualizing the distance of SNs to tumor/stroma interface<a class="anchor" aria-label="anchor" href="#visualizing-the-distance-of-sns-to-tumorstroma-interface"></a>
</h3>
<p>This plot shows the distance of each SN to the tumor/stroma
interface. Here, positive distances indicate SNs located within the
tumor region, while negative distances denote SNs within the stroma.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">obj</span>, <span class="st">"Dist2Interface"</span>, min.cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">600</span>, max.cutoff <span class="op">=</span> <span class="fl">600</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#5e3c99"</span>, high <span class="op">=</span> <span class="st">"#e66101"</span>, mid <span class="op">=</span> <span class="st">"#d9d9d9"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/umap_dist-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-spatial-ecotypes-in-the-embedding">Visualizing spatial ecotypes in the embedding<a class="anchor" aria-label="anchor" href="#visualizing-spatial-ecotypes-in-the-embedding"></a>
</h3>
<p>This plot visualizes the SEs within the spatial embedding. Each SE
represents a distinct spatial ecosystem with unique molecular and
spatial characteristics, and may also differ in cell type
composition.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"SE"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">kelly</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/umap_se-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-the-most-abundant-cell-types-in-each-sn">Visualizing the most abundant cell types in each SN<a class="anchor" aria-label="anchor" href="#visualizing-the-most-abundant-cell-types-in-each-sn"></a>
</h3>
<p>Here, we focus on visualizing the dominant cell types within each SN
to highlight the characteristic cellular composition of different
SEs.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/umap_ct-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="se-characteristics">SE characteristics<a class="anchor" aria-label="anchor" href="#se-characteristics"></a>
</h2>
<div class="section level3">
<h3 id="visualizing-ses-in-the-tissue">Visualizing SEs in the tissue<a class="anchor" aria-label="anchor" href="#visualizing-ses-in-the-tissue"></a>
</h3>
<p>The spatial distribution of SEs within the tissue can be visualized
using the <a href="../reference/SpatialView.html">SpatialView</a>
function.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">scmeta</span>, by <span class="op">=</span> <span class="st">"SE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/spatialmap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-the-cell-type-composition-of-ses">Visualizing the cell type composition of SEs<a class="anchor" aria-label="anchor" href="#visualizing-the-cell-type-composition-of-ses"></a>
</h3>
<p>The bar plot below illustrates the cell type composition within each
SE.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">CellType</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">n</span>, fill <span class="op">=</span> <span class="va">CellType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/discrete.html" class="external-link">cols25</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Cell type abundance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/bar_ct-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-the-association-between-ses-and-pre-annotated-regions">Visualizing the association between SEs and pre-annotated
regions<a class="anchor" aria-label="anchor" href="#visualizing-the-association-between-ses-and-pre-annotated-regions"></a>
</h3>
<p>This bar plot shows the enrichment of SEs in pre-defined regions
(e.g., tumor and stroma).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Region</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">n</span>, fill <span class="op">=</span> <span class="va">Region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"fill"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"qual"</span>, palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Fraction"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/bar_region-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-the-distance-of-ses-to-tumorstroma-interface">Visualizing the distance of SEs to tumor/stroma interface<a class="anchor" aria-label="anchor" href="#visualizing-the-distance-of-ses-to-tumorstroma-interface"></a>
</h3>
<p>This box plot visualizes the distribution of distances of SEs to the
tumor/stroma interface. Positive distances indicate cells located within
the tumor region, while negative distances denote cells within the
stroma. The SEs are ordered by their median distance, highlighting their
spatial localization relative to the tumor/stroma interface.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Order SEs by their distance to tumor/stroma interface</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span>  <span class="va">gg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>Mid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">Dist2Interface</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Mid</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">SE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">gg</span><span class="op">$</span><span class="va">SE</span>, levels <span class="op">=</span> <span class="va">tmp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Dist2Interface</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Distance to tumor/stroma interface (μm)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/box_dist-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="identification-of-cell-type-specific-se-markers">Identification of cell-type-specific SE markers<a class="anchor" aria-label="anchor" href="#identification-of-cell-type-specific-se-markers"></a>
</h2>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>To identify cell-type-specific SE markers, we can perform
differential expression analysis using the <a href="https://github.com/immunogenomics/presto" class="external-link"><code>presto</code>
package</a>. Below is an example of how to identify fibroblast-specific
markers for each SE.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"presto"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">{</span> </span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"immunogenomics/presto"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"presto"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ensure the normalized data is aligned with the single-cell metadata</span></span>
<span><span class="va">normdata</span> <span class="op">=</span> <span class="va">normdata</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Perform differential expression analysis for fibroblasts across SEs</span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">CellType</span><span class="op">==</span><span class="st">"Fibroblast"</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">degs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/presto/man/wilcoxauc.html" class="external-link">wilcoxauc</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">[</span>, <span class="va">idx</span><span class="op">]</span>, <span class="va">scmeta</span><span class="op">$</span><span class="va">SE</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter for significant markers with positive log-fold change</span></span>
<span><span class="va">degs</span> <span class="op">=</span> <span class="va">degs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">pval</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">degs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   feature group  avgExpr    logFC statistic       auc          pval</span></span>
<span><span class="co">## 1   ACTA2   SE6 2.474779 2.098878   4006694 0.7945178 4.327192e-195</span></span>
<span><span class="co">## 2 PLA2G2A   SE1 2.138834 1.514224   7450658 0.8350914  0.000000e+00</span></span>
<span><span class="co">## 3     PKM   SE5 2.553626 1.311940   2341716 0.8605602 2.684889e-117</span></span>
<span><span class="co">## 4 PLA2G2A   SE3 1.786796 1.115139   7220261 0.7530508 2.353682e-242</span></span>
<span><span class="co">## 5     FN1   SE4 3.050625 1.034756   4915868 0.7688155 2.080559e-150</span></span>
<span><span class="co">## 6     PKM   SE4 2.203188 1.014937   5178804 0.8099374 3.607553e-202</span></span>
<span><span class="co">##            padj   pct_in  pct_out</span></span>
<span><span class="co">## 1 5.942677e-193 72.88889 31.86990</span></span>
<span><span class="co">## 2  0.000000e+00 91.71779 34.22976</span></span>
<span><span class="co">## 3 5.530871e-115 98.56734 82.24958</span></span>
<span><span class="co">## 4 4.848585e-240 80.58865 35.54100</span></span>
<span><span class="co">## 5 2.142976e-148 98.40909 88.21910</span></span>
<span><span class="co">## 6 1.486312e-199 97.50000 81.18635</span></span></code></pre>
<p><strong>Note</strong>: The MERSCOPE data used for SE discovery
includes a limited number of genes. To enhance the analysis, you can
extend the differential expression analysis to the whole transcriptome
by aligning scRNA-seq data to the spatial transcriptomics data via
CytoSPACE.</p>
</div>
<div class="section level3">
<h3 id="visualizing-the-expression-of-cell-state-markers">Visualizing the expression of cell state markers<a class="anchor" aria-label="anchor" href="#visualizing-the-expression-of-cell-state-markers"></a>
</h3>
<p>Once you’ve identified potential markers, you can visualize their
expression across SEs. Below is an example using the ACTA2 gene.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Gene expression across all SEs</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">Expression</span> <span class="op">&lt;-</span> <span class="va">normdata</span><span class="op">[</span><span class="st">"ACTA2"</span>, <span class="va">idx</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">SE</span>, <span class="va">Expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"ACTA2 expression"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/visualizedeg-1.png" width="700"></p>
<p>You can also visualize the expression of a marker (e.g., ACTA2) in
the tissue.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Visualize gene expression in spatial</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">scmeta</span></span>
<span><span class="va">gg</span><span class="op">$</span><span class="va">Expression</span> <span class="op">&lt;-</span> <span class="va">normdata</span><span class="op">[</span><span class="st">"ACTA2"</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="../reference/SpatialView.html">SpatialView</a></span><span class="op">(</span><span class="va">gg</span>, by <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ACTA2\nexpression"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleSample_files/figure-html/viewgene-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.6.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] presto_1.0.0          Rcpp_1.0.13           SpatialEcoTyper_1.0.1</span></span>
<span><span class="co">##  [4] NMF_0.28              Biobase_2.64.0        BiocGenerics_0.50.0  </span></span>
<span><span class="co">##  [7] cluster_2.1.6         rngtools_1.5.2        registry_0.5-1       </span></span>
<span><span class="co">## [10] RANN_2.6.2            Matrix_1.7-0          R.utils_2.12.3       </span></span>
<span><span class="co">## [13] R.oo_1.26.0           R.methodsS3_1.8.2     googledrive_2.1.1    </span></span>
<span><span class="co">## [16] data.table_1.16.0     Seurat_5.1.0          SeuratObject_5.0.2   </span></span>
<span><span class="co">## [19] sp_2.1-4              ggplot2_3.5.1         dplyr_1.1.4          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">##   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">##   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4             doParallel_1.0.17          </span></span>
<span><span class="co">##   [9] pals_1.9                    globals_0.16.3             </span></span>
<span><span class="co">##  [11] lattice_0.22-6              MASS_7.3-60.2              </span></span>
<span><span class="co">##  [13] magrittr_2.0.3              plotly_4.10.4              </span></span>
<span><span class="co">##  [15] sass_0.4.9                  rmarkdown_2.28             </span></span>
<span><span class="co">##  [17] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span><span class="co">##  [19] httpuv_1.6.15               glmGamPoi_1.16.0           </span></span>
<span><span class="co">##  [21] sctransform_0.4.1           spam_2.10-0                </span></span>
<span><span class="co">##  [23] spatstat.sparse_3.1-0       reticulate_1.39.0          </span></span>
<span><span class="co">##  [25] mapproj_1.2.11              cowplot_1.1.3              </span></span>
<span><span class="co">##  [27] pbapply_1.7-2               RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [29] maps_3.4.2                  zlibbioc_1.50.0            </span></span>
<span><span class="co">##  [31] abind_1.4-5                 GenomicRanges_1.56.1       </span></span>
<span><span class="co">##  [33] Rtsne_0.17                  purrr_1.0.2                </span></span>
<span><span class="co">##  [35] GenomeInfoDbData_1.2.12     circlize_0.4.16            </span></span>
<span><span class="co">##  [37] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">##  [39] ggrepel_0.9.6               irlba_2.3.5.1              </span></span>
<span><span class="co">##  [41] listenv_0.9.1               spatstat.utils_3.1-0       </span></span>
<span><span class="co">##  [43] goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span><span class="co">##  [45] spatstat.random_3.3-1       fitdistrplus_1.2-1         </span></span>
<span><span class="co">##  [47] parallelly_1.38.0           DelayedMatrixStats_1.26.0  </span></span>
<span><span class="co">##  [49] pkgdown_2.1.0               DelayedArray_0.30.1        </span></span>
<span><span class="co">##  [51] leiden_0.4.3.1              codetools_0.2-20           </span></span>
<span><span class="co">##  [53] tidyselect_1.2.1            shape_1.4.6.1              </span></span>
<span><span class="co">##  [55] farver_2.1.2                UCSC.utils_1.0.0           </span></span>
<span><span class="co">##  [57] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">##  [59] spatstat.explore_3.3-2      jsonlite_1.8.8             </span></span>
<span><span class="co">##  [61] GetoptLong_1.0.5            progressr_0.14.0           </span></span>
<span><span class="co">##  [63] ggridges_0.5.6              survival_3.6-4             </span></span>
<span><span class="co">##  [65] iterators_1.0.14            systemfonts_1.1.0          </span></span>
<span><span class="co">##  [67] foreach_1.5.2               tools_4.4.1                </span></span>
<span><span class="co">##  [69] ragg_1.3.2                  ica_1.0-3                  </span></span>
<span><span class="co">##  [71] glue_1.7.0                  SparseArray_1.4.8          </span></span>
<span><span class="co">##  [73] gridExtra_2.3               xfun_0.52                  </span></span>
<span><span class="co">##  [75] MatrixGenerics_1.16.0       GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">##  [77] withr_3.0.1                 BiocManager_1.30.25        </span></span>
<span><span class="co">##  [79] fastmap_1.2.0               fansi_1.0.6                </span></span>
<span><span class="co">##  [81] digest_0.6.37               R6_2.5.1                   </span></span>
<span><span class="co">##  [83] mime_0.12                   textshaping_0.4.0          </span></span>
<span><span class="co">##  [85] colorspace_2.1-1            scattermore_1.2            </span></span>
<span><span class="co">##  [87] tensor_1.5                  dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [89] spatstat.data_3.1-2         utf8_1.2.4                 </span></span>
<span><span class="co">##  [91] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">##  [93] S4Arrays_1.4.1              httr_1.4.7                 </span></span>
<span><span class="co">##  [95] htmlwidgets_1.6.4           uwot_0.2.2                 </span></span>
<span><span class="co">##  [97] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">##  [99] ComplexHeatmap_2.20.0       lmtest_0.9-40              </span></span>
<span><span class="co">## [101] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">## [103] dotCall64_1.1-1             clue_0.3-65                </span></span>
<span><span class="co">## [105] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">## [107] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">## [109] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">## [111] rjson_0.2.22                nlme_3.1-164               </span></span>
<span><span class="co">## [113] curl_5.2.2                  cachem_1.1.0               </span></span>
<span><span class="co">## [115] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">## [117] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">## [119] miniUI_0.1.1.1              desc_1.4.3                 </span></span>
<span><span class="co">## [121] pillar_1.9.0                grid_4.4.1                 </span></span>
<span><span class="co">## [123] vctrs_0.6.5                 promises_1.3.0             </span></span>
<span><span class="co">## [125] xtable_1.8-4                evaluate_0.24.0            </span></span>
<span><span class="co">## [127] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">## [129] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [131] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">## [133] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">## [135] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span><span class="co">## [137] deldir_2.0-4                gridBase_0.4-7             </span></span>
<span><span class="co">## [139] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">## [141] spatstat.geom_3.3-2         RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [143] patchwork_1.2.0             sparseMatrixStats_1.16.0   </span></span>
<span><span class="co">## [145] future_1.34.0               shiny_1.9.1                </span></span>
<span><span class="co">## [147] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [149] ROCR_1.0-11                 gargle_1.5.2               </span></span>
<span><span class="co">## [151] igraph_2.0.3                bslib_0.8.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
