<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Development of NMF Models for Spatial Ecotype Recovery â€¢ SpatialEcoTyper</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Development of NMF Models for Spatial Ecotype Recovery">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Development of NMF Models for Spatial Ecotype Recovery</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/vignettes/TrainRecoveryModels.Rmd" class="external-link"><code>vignettes/TrainRecoveryModels.Rmd</code></a></small>
      <div class="d-none name"><code>TrainRecoveryModels.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>After performing Spatial EcoTyper analysis, users can develop machine
learning models for spatial ecotype (SE) recovery, including
reference-guided SE annotation in held-out single-cell datasets
(scRNA-seq or single-cell spatial transcriptomics) and SE deconvolution
to infer SE abundances from bulk gene expression data (bulk RNA-seq or
Visium). The Spatial EcoTyper framework employs non-negative matrix
factorization (NMF) as the core algorithm in both analyses, given its
robustness and proven effectiveness in recovering carcinoma ecotypes
from multimodal expression data in our <a href="https://www.sciencedirect.com/science/article/pii/S0092867421010618" class="external-link">prior
study</a>. Here, we will demonstrate how to train NMF models for SE
recovery from single-cell gene expression data and for SE deconvolution
from bulk gene expression data separately.</p>
</div>
<div class="section level2">
<h2 id="development-of-recovery-model-to-annotate-ses-in-single-cell-gene-expression-data">Development of recovery model to annotate SEs in single-cell gene
expression data<a class="anchor" aria-label="anchor" href="#development-of-recovery-model-to-annotate-ses-in-single-cell-gene-expression-data"></a>
</h2>
<p>In this tutorial, we will use the Vizgen MERSCOPE data of two samples
to demonstrate the process for developing cell-type-specific NMF models
for recovering SEs from single-cell gene expression profiles, including
both single-cell spatial transcriptomics and single-cell RNA-seq
data.</p>
<p>The gene expression matrices, cell type and SE annotations required
for the training can be obtained from <a href="https://drive.google.com/open?id=17lFcu0DrUChy_1Edfs-dvVAPQLenL-jT&amp;usp=drive_fs" class="external-link">here</a>.</p>
<p><strong>First load required packages for this vignette</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p><strong>Download the data from Google Drive</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17mbc56M0GTq-9MR0GqHas-L-yIXdnxtu"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17nL0ceONPw8ByAtDe34DiXiGLMVBLlPX"</span><span class="op">)</span>, <span class="st">"HumanColonCancerPatient2_subset_counts.tsv.gz"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"1t3MX84RHm5BQMzcAvXtmzc-Dj4e6Y7Ik"</span><span class="op">)</span>, <span class="st">"MultiSE_metadata_final.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Load gene expression matrix</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Read the expression matrix for the first sample</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>, </span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co">## set the first column as row names</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="co">## drop the first column</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>  <span class="co">## Normalize the expression matrix</span></span>
<span></span>
<span><span class="co">## Read the expression matrix for the second sample</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanColonCancerPatient2_subset_counts.tsv.gz"</span>, </span>
<span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata2</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co">## set the first column as row names</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="va">scdata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="co">## drop the first column</span></span>
<span><span class="va">scdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span>  <span class="co">## Normalize the expression matrix</span></span>
<span></span>
<span><span class="co">## Combine the two expression matrices</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span>, <span class="va">scdata2</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657</span></span>
<span><span class="co">## PDK4                             0.000000                         4.221184</span></span>
<span><span class="co">## TNFRSF17                         0.000000                         0.000000</span></span>
<span><span class="co">## ICAM3                            0.000000                         0.000000</span></span>
<span><span class="co">## FAP                              3.552438                         0.000000</span></span>
<span><span class="co">## GZMB                             0.000000                         0.000000</span></span>
<span><span class="co">## TSC2                             0.000000                         0.000000</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660</span></span>
<span><span class="co">## PDK4                             3.208945                                0</span></span>
<span><span class="co">## TNFRSF17                         0.000000                                0</span></span>
<span><span class="co">## ICAM3                            0.000000                                0</span></span>
<span><span class="co">## FAP                              0.000000                                0</span></span>
<span><span class="co">## GZMB                             0.000000                                0</span></span>
<span><span class="co">## TSC2                             0.000000                                0</span></span>
<span><span class="co">##          HumanMelanomaPatient1__cell_3661</span></span>
<span><span class="co">## PDK4                                    0</span></span>
<span><span class="co">## TNFRSF17                                0</span></span>
<span><span class="co">## ICAM3                                   0</span></span>
<span><span class="co">## FAP                                     0</span></span>
<span><span class="co">## GZMB                                    0</span></span>
<span><span class="co">## TSC2                                    0</span></span></code></pre>
<p><strong>Load cell type and SE annotations</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load the single-cell meta data derived from MultiSpatialEcoTyper analysis</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"MultiSE_metadata_final.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                     header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">scmeta</span><span class="op">$</span><span class="va">SE</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">$</span><span class="va">SE</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="st">"nonSE"</span> <span class="co">## cells with NA predictions were assigned into the nonSE group</span></span>
<span></span>
<span><span class="co">## Matches the row names of meta data and column names of expression matrix</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="va">cells</span><span class="op">]</span></span>
<span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">scmeta</span><span class="op">[</span><span class="va">cells</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"CellType"</span>, <span class="st">"SE"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  Sample   CellType  SE</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3655   SKCM Fibroblast SE6</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3657   SKCM Fibroblast SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3658   SKCM Fibroblast SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3660   SKCM Fibroblast SE1</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3661   SKCM Fibroblast SE6</span></span>
<span><span class="co">## HumanMelanomaPatient1__cell_3663   SKCM Fibroblast SE6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="training">Training<a class="anchor" aria-label="anchor" href="#training"></a>
</h3>
<p>The <a href="../reference/NMFGenerateWList.html">NMFGenerateWList</a>
function will be used to train cell-type specific NMF models for
recovering SEs. Prior to NMF, each gene is scaled to mean 0 and unit
variance within each cell type-specific expression matrix. To satisfy
the non-negativity requirement of NMF, the cell type-specific expression
matrices are individually processed using posneg transformation, which
converts each expression matrix into two matrices, one containing only
positive values and the other containing only negative values with the
sign inverted. The two matrices are subsequently concatenated to produce
the training data. The <a href="../reference/NMFGenerateWList.html">NMFGenerateWList</a> function
will return a list of <code>feature Ã— SE</code> matrices, each
representing a cell type-specific gene signature matrix with rows as
features and columns as SEs.</p>
<p><strong>Note</strong>: By default, the <a href="../reference/NMFGenerateWList.html">NMFGenerateWList</a> selects
the top 2000 variable features (<code>nfeature</code> argument) and
down-samples 2500 cells (<code>downsample</code> argument) within each
cell type for training. After training, it retains 50 genes per SE
(<code>nfeature.per.se</code> argument) in the model. Users can tune the
parameters accordingly.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># training</span></span>
<span><span class="va">Ws</span> <span class="op">=</span> <span class="fu"><a href="../reference/NMFGenerateWList.html">NMFGenerateWList</a></span><span class="op">(</span><span class="va">scdata</span>, <span class="va">scmeta</span>, </span>
<span>                      CellType <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>                      SE <span class="op">=</span> <span class="st">"SE"</span>, Sample <span class="op">=</span> <span class="st">"Sample"</span><span class="op">)</span></span>
<span><span class="co">## This step would take ~3 mins for the demo.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "B"           "CD4T"        "CD8T"        "DC"          "Endothelial"</span></span>
<span><span class="co">## [6] "Fibroblast"  "Macrophage"  "NK"          "Plasma"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   SE6      SE10       SE8       SE5       SE4       SE3</span></span>
<span><span class="co">## CCL5__neg   0.5145161 0.5397050 0.4430638 0.7122035 0.5261429 0.4446076</span></span>
<span><span class="co">## XBP1__neg   0.4026504 0.5881324 0.4534266 0.9834871 0.4475261 0.3841000</span></span>
<span><span class="co">## NFKBIA__neg 0.5205229 0.3957970 0.3920701 0.3677413 0.6758230 0.5218947</span></span>
<span><span class="co">## CDKN1B__neg 0.3609662 0.5577056 0.5121339 0.7531579 0.4968649 0.4651990</span></span>
<span><span class="co">## CXCR6__pos  0.4714863 0.4993970 0.2598859 0.8154450 0.6067278 0.4657444</span></span>
<span><span class="co">## NFKB1__neg  0.5238008 0.4750225 0.4536049 0.4313634 0.6072406 0.5600733</span></span>
<span><span class="co">##                   SE1       SE2       SE7       SE9</span></span>
<span><span class="co">## CCL5__neg   0.6530544 0.5411019 0.7678666 0.6780336</span></span>
<span><span class="co">## XBP1__neg   0.6401507 0.4460380 0.5866628 0.8487692</span></span>
<span><span class="co">## NFKBIA__neg 0.6758407 0.5994373 0.5995011 1.0137611</span></span>
<span><span class="co">## CDKN1B__neg 0.7280766 0.4544132 0.6294411 0.7415819</span></span>
<span><span class="co">## CXCR6__pos  0.6089321 0.6168503 0.4655022 0.8689492</span></span>
<span><span class="co">## NFKB1__neg  0.5656818 0.4970913 0.6098723 0.8485226</span></span></code></pre>
<p>The resulting W matrices could be used to recover SEs from spatial
transcriptomics and single-cell RNA-seq using the <a href="../reference/RecoverSE.html">RecoverSE</a> function. The detailed
documentation for SE recovery is available at <a href="Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell
Gene Expression Data</a> and <a href="Recovery_Spatial.html">Recovery of
Spatial Ecotypes from Spatial Transcriptomics Data</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="development-of-se-deconvolution-model-to-infer-se-abundances-from-bulk-gene-expression-data">Development of SE deconvolution model to infer SE abundances from
bulk gene expression data<a class="anchor" aria-label="anchor" href="#development-of-se-deconvolution-model-to-infer-se-abundances-from-bulk-gene-expression-data"></a>
</h2>
<p>In this tutorial, we will demonstrate how to develop an NMF model to
deconvolve SEs from bulk gene expression data. This process requires
bulk training data with known SE compositions, which we can generated by
aggregating single-cell RNA-seq data into pseudo-bulk mixtures. For
demonstration, we will use single-cell RNA-seq data from a melanoma
sample.</p>
<p>The raw count for this sample is available at <a href="https://drive.google.com/open?id=17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei&amp;usp=drive_fs" class="external-link">WU2161_counts.tsv</a>
and SE recovery results are available at <a href="https://drive.google.com/open?id=17kFjOHhmWxWAKm-LDqy6wV27o5lvz8L2&amp;usp=drive_fs" class="external-link">WU2161_RecoveredSEs.tsv</a>.</p>
<div class="section level3">
<h3 id="loading-data-1">Loading data<a class="anchor" aria-label="anchor" href="#loading-data-1"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download single-cell gene expression matrix and SE recovery results.</span></span>
<span><span class="co"># The downloads should be finished within 1min.</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17VAeOnz6vTt2s0ZeTrK1kITdJ3Yus4ei"</span><span class="op">)</span>, <span class="st">"WU2161_counts.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"17kFjOHhmWxWAKm-LDqy6wV27o5lvz8L2"</span><span class="op">)</span>, <span class="st">"WU2161_RecoveredSEs.tsv"</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load single-cell gene expression matrix.</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"WU2161_counts.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="co">## Set the first column as row names</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">## Drop the first column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            AAACCTGCACATGACT-1 AAACCTGGTGGTCCGT-1 AAACCTGGTTTGTGTG-1</span></span>
<span><span class="co">## AL627309.1                  0                  0                  0</span></span>
<span><span class="co">## AL627309.5                  0                  0                  0</span></span>
<span><span class="co">## AP006222.2                  0                  0                  0</span></span>
<span><span class="co">## AC114498.1                  0                  0                  0</span></span>
<span><span class="co">## AL669831.2                  0                  0                  0</span></span>
<span><span class="co">## LINC01409                   0                  0                  0</span></span>
<span><span class="co">##            AAACCTGTCCGATATG-1 AAACCTGTCTAACGGT-1</span></span>
<span><span class="co">## AL627309.1                  0                  0</span></span>
<span><span class="co">## AL627309.5                  1                  0</span></span>
<span><span class="co">## AP006222.2                  0                  0</span></span>
<span><span class="co">## AC114498.1                  0                  0</span></span>
<span><span class="co">## AL669831.2                  0                  0</span></span>
<span><span class="co">## LINC01409                   0                  0</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load SE recovery results</span></span>
<span><span class="va">SEs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"WU2161_RecoveredSEs.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="fl">1</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">SEs</span> <span class="op">=</span> <span class="va">SEs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">]</span> <span class="co"># extract SE predictions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1337</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">SEs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## SEs</span></span>
<span><span class="co">## nonSE   SE1   SE2   SE3   SE4   SE5   SE6   SE7   SE8   SE9 </span></span>
<span><span class="co">##   912    71    45    24     5    43    74    75    71    17</span></span></code></pre>
<p><strong>Note</strong>: For demonstration purposes, we used 1,337
cells grouped into SEs to create pseudo-bulk mixtures. While this
limited cell number offers a basic example, it does not fully capture
the diverse characteristics of SEs, potentially affecting the robustness
of model training and subsequent SE recovery. In practice, we recommend
using a more comprehensive dataset that accurately reflects the
properties of SEs, ensuring that the training process results in a
reliable recovery model.</p>
</div>
<div class="section level3">
<h3 id="creating-pseudobulk-data">Creating pseudobulk data<a class="anchor" aria-label="anchor" href="#creating-pseudobulk-data"></a>
</h3>
<p>The <a href="../reference/CreatePseudobulks.html">CreatePseudobulks</a>
function will be used to create pseudo-bulk mixtures from single-cell
(spatial) transcriptomics data with all cells grouped into SEs. It
samples cell fractions from a Gaussian distribution, sets negative
fractions to 0 and re-normalizes fractions to sum to 1 across all SEs.
Based on the resulting fractions, it samples 1,000 cells per pseudo-bulk
mixture with replacement, aggregates their transcriptomes in non-log
linear space, and normalize the resulting mixture to logarithm CPM using
Seuratâ€™s <code>NormalizeData</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">=</span> <span class="fu"><a href="../reference/CreatePseudobulks.html">CreatePseudobulks</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>, groups <span class="op">=</span> <span class="va">SEs</span>, n_mixtures <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## This step took ~4 min for the demo.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Mixtures</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span> <span class="co">## Gene expression matrix of pseudobulks</span></span></code></pre></div>
<pre><code><span><span class="co">##            Pseudobulk1  Pseudobulk2 Pseudobulk3  Pseudobulk4 Pseudobulk5</span></span>
<span><span class="co">## AL627309.1 0.034510811 0.0248976190 0.017730486 0.0116095624 0.012006108</span></span>
<span><span class="co">## AL627309.5 0.048433430 0.0423069886 0.037175210 0.0388799578 0.053369352</span></span>
<span><span class="co">## AP006222.2 0.001979247 0.0017176501 0.002230716 0.0014918933 0.003404467</span></span>
<span><span class="co">## AC114498.1 0.002768040 0.0009226095 0.000000000 0.0009244557 0.000000000</span></span>
<span><span class="co">## AL669831.2 0.000000000 0.0000000000 0.000000000 0.0000000000 0.000000000</span></span>
<span><span class="co">## LINC01409  0.049101136 0.0454449643 0.039554865 0.0454409663 0.040022663</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Fracs</span><span class="op">)</span> <span class="co">## SE fractions in pseudobulks</span></span></code></pre></div>
<pre><code><span><span class="co">##                  nonSE        SE1        SE2        SE3         SE4        SE5</span></span>
<span><span class="co">## Pseudobulk1 0.13566259 0.03714595 0.08700476 0.11247413 0.006952763 0.16925317</span></span>
<span><span class="co">## Pseudobulk2 0.11053902 0.09494654 0.07522259 0.13304700 0.035826405 0.09667114</span></span>
<span><span class="co">## Pseudobulk3 0.09117524 0.10499265 0.09641997 0.08563370 0.131132847 0.14371435</span></span>
<span><span class="co">## Pseudobulk4 0.14938164 0.05010742 0.06586831 0.11229859 0.086042379 0.12967002</span></span>
<span><span class="co">## Pseudobulk5 0.16958606 0.04087059 0.07243148 0.12060727 0.035014712 0.12229811</span></span>
<span><span class="co">## Pseudobulk6 0.15181360 0.05296392 0.14561045 0.05306015 0.117005773 0.06958625</span></span>
<span><span class="co">##                    SE6        SE7        SE8        SE9</span></span>
<span><span class="co">## Pseudobulk1 0.13480120 0.04463829 0.10635447 0.16571267</span></span>
<span><span class="co">## Pseudobulk2 0.10409005 0.10669564 0.15664085 0.08632077</span></span>
<span><span class="co">## Pseudobulk3 0.06665823 0.04219794 0.08302129 0.15505378</span></span>
<span><span class="co">## Pseudobulk4 0.14790042 0.10413103 0.05321619 0.10138399</span></span>
<span><span class="co">## Pseudobulk5 0.12102642 0.14975237 0.04712587 0.12128714</span></span>
<span><span class="co">## Pseudobulk6 0.11155113 0.12587378 0.06088209 0.11165287</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="training-1">Training<a class="anchor" aria-label="anchor" href="#training-1"></a>
</h3>
<p>The <a href="../reference/NMFGenerateW.html">NMFGenerateW</a>
function will be used to train an NMF model for SE deconvolution based
on the provided SE fractions and gene expression matrix. Before applying
NMF, each geneâ€™s expression is scaled to have a mean of 0 and unit
variance by default. To meet the non-negativity requirement of NMF, the
expression matrix is transformed using the posneg method. This
transformation splits the expression matrix into two matrices: one
containing only positive values and the other containing the absolute
values of the negative values. These two matrices are then concatenated
to form the final training data for the NMF model.</p>
<p><strong>Note</strong>: By default, the <a href="../reference/NMFGenerateW.html">NMFGenerateW</a> selects the top
2000 variable features (<code>nfeature</code> argument) for training.
After training, it retains 50 genes per SE (<code>nfeature.per.se</code>
argument) in the model. Users can tune the parameters accordingly.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">W</span> <span class="op">=</span> <span class="fu"><a href="../reference/NMFGenerateW.html">NMFGenerateW</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">Fracs</span>, <span class="va">result</span><span class="op">$</span><span class="va">Mixtures</span><span class="op">)</span></span>
<span><span class="co">## This step took less than 1 min for the demo.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      nonSE          SE1          SE2          SE3          SE4</span></span>
<span><span class="co">## IGHV3-49__pos 2.220446e-16 2.220446e-16 1.176118e-15 2.220446e-16 2.220446e-16</span></span>
<span><span class="co">## IGHG2__pos    2.220446e-16 2.220446e-16 2.220446e-16 4.371307e-12 2.220446e-16</span></span>
<span><span class="co">## IGKV2-28__pos 2.220446e-16 2.220446e-16 2.220446e-16 2.220446e-16 2.220446e-16</span></span>
<span><span class="co">## IGHV1-2__neg  3.516628e-15 4.123817e-01 9.560283e-02 5.087919e-01 1.835669e+00</span></span>
<span><span class="co">## IGHV3-49__neg 3.818675e-05 1.055389e+00 2.649221e-13 1.820891e-01 4.904564e-05</span></span>
<span><span class="co">## IGHV5-51__pos 2.220446e-16 2.220446e-16 2.220446e-16 2.220446e-16 2.220446e-16</span></span>
<span><span class="co">##                        SE5          SE6          SE7          SE8          SE9</span></span>
<span><span class="co">## IGHV3-49__pos 2.220446e-16 1.390739e-13 4.909769e+00 1.867123e-15 2.220446e-16</span></span>
<span><span class="co">## IGHG2__pos    2.220446e-16 7.146394e-05 4.572860e+00 2.220446e-16 2.220446e-16</span></span>
<span><span class="co">## IGKV2-28__pos 2.220446e-16 2.220446e-16 4.542140e+00 1.800026e-12 2.220446e-16</span></span>
<span><span class="co">## IGHV1-2__neg  2.811440e-01 3.454982e-03 1.250619e+00 1.488007e-01 1.197966e-06</span></span>
<span><span class="co">## IGHV3-49__neg 7.453432e-07 4.681541e-02 2.220446e-16 1.442175e-01 3.107528e+00</span></span>
<span><span class="co">## IGHV5-51__pos 2.220446e-16 2.220446e-16 4.513844e+00 2.220446e-16 2.220446e-16</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Save the model for late use.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">W</span>, <span class="st">"SE_Deconvolution_W.rds"</span><span class="op">)</span></span></code></pre></div>
<p>The resulting W matrix can be used to infer SE abundances from bulk
gene expression profiles using the <a href="../reference/DeconvoluteSE.html">DeconvoluteSE</a> function. The
detailed documentation for SE deconvolution is available at <a href="Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene
Expression Data</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>The session info allows users to replicate the exact environment and
identify potential discrepancies in package versions or configurations
that might be causing problems.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS 15.6.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SpatialEcoTyper_1.0.1 RANN_2.6.2            Matrix_1.7-0         </span></span>
<span><span class="co">##  [4] dplyr_1.1.4           data.table_1.16.0     NMF_0.28             </span></span>
<span><span class="co">##  [7] Biobase_2.64.0        BiocGenerics_0.50.0   cluster_2.1.6        </span></span>
<span><span class="co">## [10] rngtools_1.5.2        registry_0.5-1        Seurat_5.1.0         </span></span>
<span><span class="co">## [13] SeuratObject_5.0.2    sp_2.1-4              googledrive_2.1.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     shape_1.4.6.1          rstudioapi_0.16.0     </span></span>
<span><span class="co">##   [4] jsonlite_1.8.8         magrittr_2.0.3         spatstat.utils_3.1-0  </span></span>
<span><span class="co">##   [7] rmarkdown_2.28         GlobalOptions_0.1.2    fs_1.6.4              </span></span>
<span><span class="co">##  [10] ragg_1.3.2             vctrs_0.6.5            ROCR_1.0-11           </span></span>
<span><span class="co">##  [13] spatstat.explore_3.3-2 htmltools_0.5.8.1      curl_5.2.2            </span></span>
<span><span class="co">##  [16] sass_0.4.9             sctransform_0.4.1      parallelly_1.38.0     </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-24     bslib_0.8.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">##  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">##  [25] plotly_4.10.4          zoo_1.8-12             cachem_1.1.0          </span></span>
<span><span class="co">##  [28] igraph_2.0.3           iterators_1.0.14       mime_0.12             </span></span>
<span><span class="co">##  [31] lifecycle_1.0.4        pkgconfig_2.0.3        R6_2.5.1              </span></span>
<span><span class="co">##  [34] fastmap_1.2.0          clue_0.3-65            fitdistrplus_1.2-1    </span></span>
<span><span class="co">##  [37] future_1.34.0          shiny_1.9.1            digest_0.6.37         </span></span>
<span><span class="co">##  [40] colorspace_2.1-1       S4Vectors_0.42.1       patchwork_1.2.0       </span></span>
<span><span class="co">##  [43] tensor_1.5             RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">##  [46] textshaping_0.4.0      progressr_0.14.0       fansi_1.0.6           </span></span>
<span><span class="co">##  [49] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       </span></span>
<span><span class="co">##  [52] abind_1.4-5            compiler_4.4.1         gargle_1.5.2          </span></span>
<span><span class="co">##  [55] withr_3.0.1            doParallel_1.0.17      fastDummies_1.7.4     </span></span>
<span><span class="co">##  [58] R.utils_2.12.3         MASS_7.3-60.2          rjson_0.2.22          </span></span>
<span><span class="co">##  [61] tools_4.4.1            lmtest_0.9-40          httpuv_1.6.15         </span></span>
<span><span class="co">##  [64] future.apply_1.11.2    goftest_1.2-3          R.oo_1.26.0           </span></span>
<span><span class="co">##  [67] glue_1.7.0             nlme_3.1-164           promises_1.3.0        </span></span>
<span><span class="co">##  [70] grid_4.4.1             gridBase_0.4-7         Rtsne_0.17            </span></span>
<span><span class="co">##  [73] reshape2_1.4.4         generics_0.1.3         gtable_0.3.5          </span></span>
<span><span class="co">##  [76] spatstat.data_3.1-2    R.methodsS3_1.8.2      tidyr_1.3.1           </span></span>
<span><span class="co">##  [79] utf8_1.2.4             spatstat.geom_3.3-2    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">##  [82] foreach_1.5.2          ggrepel_0.9.6          pillar_1.9.0          </span></span>
<span><span class="co">##  [85] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">##  [88] later_1.3.2            circlize_0.4.16        splines_4.4.1         </span></span>
<span><span class="co">##  [91] lattice_0.22-6         survival_3.6-4         deldir_2.0-4          </span></span>
<span><span class="co">##  [94] tidyselect_1.2.1       ComplexHeatmap_2.20.0  miniUI_0.1.1.1        </span></span>
<span><span class="co">##  [97] pbapply_1.7-2          knitr_1.48             gridExtra_2.3         </span></span>
<span><span class="co">## [100] IRanges_2.38.1         scattermore_1.2        stats4_4.4.1          </span></span>
<span><span class="co">## [103] xfun_0.52              matrixStats_1.4.1      stringi_1.8.4         </span></span>
<span><span class="co">## [106] lazyeval_0.2.2         yaml_2.3.10            evaluate_0.24.0       </span></span>
<span><span class="co">## [109] codetools_0.2-20       tibble_3.2.1           BiocManager_1.30.25   </span></span>
<span><span class="co">## [112] cli_3.6.3              uwot_0.2.2             xtable_1.8-4          </span></span>
<span><span class="co">## [115] reticulate_1.39.0      systemfonts_1.1.0      munsell_0.5.1         </span></span>
<span><span class="co">## [118] jquerylib_0.1.4        Rcpp_1.0.13            globals_0.16.3        </span></span>
<span><span class="co">## [121] spatstat.random_3.3-1  png_0.1-8              spatstat.univar_3.0-1 </span></span>
<span><span class="co">## [124] pkgdown_2.1.0          ggplot2_3.5.1          dotCall64_1.1-1       </span></span>
<span><span class="co">## [127] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          </span></span>
<span><span class="co">## [130] ggridges_0.5.6         crayon_1.5.3           leiden_0.4.3.1        </span></span>
<span><span class="co">## [133] purrr_1.0.2            GetoptLong_1.0.5       rlang_1.1.4           </span></span>
<span><span class="co">## [136] cowplot_1.1.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
