<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ SpatialEcoTyper</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="discovery-of-spatial-ecotypes-ses-">Discovery of spatial ecotypes (SEs)<a class="anchor" aria-label="anchor" href="#discovery-of-spatial-ecotypes-ses-"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="SpatialEcoTyper.html">SpatialEcoTyper()</a></code>

        </dt>
        <dd>Identify Spatial EcoTypes from Single-cell Spatial Data (A Single Sample)</dd>
      </dl><dl><dt>

          <code><a href="MultiSpatialEcoTyper.html">MultiSpatialEcoTyper()</a></code>

        </dt>
        <dd>Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes</dd>
      </dl><dl><dt>

          <code><a href="IntegrateSpatialEcoTyper.html">IntegrateSpatialEcoTyper()</a></code>

        </dt>
        <dd>Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes</dd>
      </dl><dl><dt>

          <code><a href="GetSpatialMetacells.html">GetSpatialMetacells()</a></code>

        </dt>
        <dd>Construct Spatial Metacells from Single-Cell Spatial Data</dd>
      </dl><dl><dt>

          <code><a href="ComputeFCs.html">ComputeFCs()</a></code>

        </dt>
        <dd>Compute Cell-Type-Specific Fold Changes (FCs) for Spatial Clusters</dd>
      </dl><dl><dt>

          <code><a href="SNF2.html">SNF2()</a></code>

        </dt>
        <dd>Enhanced Similarity Network Fusion</dd>
      </dl><dl><dt>

          <code><a href="GetPCList.html">GetPCList()</a></code>

        </dt>
        <dd>Generate Principal Component (PC) List for Spatial Neighborhoods</dd>
      </dl><dl><dt>

          <code><a href="getSN.html">getSN()</a></code>

        </dt>
        <dd>Construct Similarity Network</dd>
      </dl><dl><dt>

          <code><a href="GetSNList.html">GetSNList()</a></code>

        </dt>
        <dd>Construct Cell-Type-Specific Similarity Network</dd>
      </dl><dl><dt>

          <code><a href="Integrate.html">Integrate()</a></code>

        </dt>
        <dd>Integrate Spatial Clusters From Multiple Samples Via Similarity Network Fusion</dd>
      </dl></div><div class="section level2">
      <h2 id="development-of-se-recovery-models">Development of SE recovery models<a class="anchor" aria-label="anchor" href="#development-of-se-recovery-models"></a></h2>

      <div class="section-desc"><p>Developing models for recovering spatial ecotypes from single-cell spatial data, scRNA-seq data, Visium spatial data, or bulk tumor expression data.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="CreatePseudobulks.html">CreatePseudobulks()</a></code>

        </dt>
        <dd>Create Pseudo-bulk Mixtures</dd>
      </dl><dl><dt>

          <code><a href="NMFGenerateW.html">NMFGenerateW()</a></code>

        </dt>
        <dd>Train SE Deconvolution Model</dd>
      </dl><dl><dt>

          <code><a href="NMFGenerateWList.html">NMFGenerateWList()</a></code>

        </dt>
        <dd>Train Cell Type-Specific NMF Models for Recovering Spatial EcoTypes</dd>
      </dl></div><div class="section level2">
      <h2 id="se-recovery">SE recovery<a class="anchor" aria-label="anchor" href="#se-recovery"></a></h2>

      <div class="section-desc"><p>Recovering spatial ecotypes from single-cell spatial transcriptomics, scRNA-seq, Visium, or bulk tumor expression data.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="RecoverSE.html">RecoverSE()</a></code>

        </dt>
        <dd>Recovery of SEs Using Pretrained NMF Models</dd>
      </dl><dl><dt>

          <code><a href="DeconvoluteSE.html">DeconvoluteSE()</a></code>

        </dt>
        <dd>Infer SE Abundances Using a Pretrained NMF Model</dd>
      </dl><dl><dt>

          <code><a href="NMFpredict.html">NMFpredict()</a></code>

        </dt>
        <dd>Prediction Using Pretrained NMF Model</dd>
      </dl></div><div class="section level2">
      <h2 id="nmf-non-negative-matrix-factorization-clustering">NMF (Non-negative Matrix Factorization) clustering<a class="anchor" aria-label="anchor" href="#nmf-non-negative-matrix-factorization-clustering"></a></h2>

      <div class="section-desc"><p>Clustering via NMF.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="nmfClustering.html">nmfClustering()</a></code>

        </dt>
        <dd>Robust Clustering via NMF (non-negative matrix factorization)</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="SpatialView.html">SpatialView()</a></code>

        </dt>
        <dd>Visualize Spatial Landscape of Cells / Spots</dd>
      </dl><dl><dt>

          <code><a href="HeatmapView.html">HeatmapView()</a></code>

        </dt>
        <dd>Draw Heatmap</dd>
      </dl><dl><dt>

          <code><a href="drawRectangleAnnotation.html">drawRectangleAnnotation()</a></code>

        </dt>
        <dd>Draw Rectangle Annotations</dd>
      </dl><dl><dt>

          <code><a href="getColors.html">getColors()</a></code>

        </dt>
        <dd>Generate a List of Colors</dd>
      </dl></div><div class="section level2">
      <h2 id="data-processing">Data processing<a class="anchor" aria-label="anchor" href="#data-processing"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="PreprocessST.html">PreprocessST()</a></code>

        </dt>
        <dd>Preprocess Spatial Transcriptomics Data</dd>
      </dl><dl><dt>

          <code><a href="AnnotateCells.html">AnnotateCells()</a></code>

        </dt>
        <dd>Extract Spatial Ecotype Annotations for Single Cells</dd>
      </dl><dl><dt>

          <code><a href="Znorm.html">Znorm()</a></code>

        </dt>
        <dd>Weighted / Unweighted Uni-variance Normalization</dd>
      </dl><dl><dt>

          <code><a href="rankSparse.html">rankSparse()</a></code>

        </dt>
        <dd>Transform a Sparse Matrix to Rank Space (Rank Non-zeros in Each Column)</dd>
      </dl><dl><dt>

          <code><a href="matrixMultiply.html">matrixMultiply()</a></code>

        </dt>
        <dd>Matrix Multiplication with Minibatching and Parallel Processing</dd>
      </dl><dl><dt>

          <code><a href="fillspots.html">fillspots()</a></code>

        </dt>
        <dd>Handle Missing Values</dd>
      </dl><dl><dt>

          <code><a href="mostFrequent.html">mostFrequent()</a></code>

        </dt>
        <dd>Identify the most frequent category in a vector</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

