<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes — IntegrateSpatialEcoTyper • SpatialEcoTyper</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes — IntegrateSpatialEcoTyper"><meta name="description" content="Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes"><meta property="og:description" content="Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes"><meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes</h1>
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/R/IntegrateSpatialEcoTyper.R" class="external-link"><code>R/IntegrateSpatialEcoTyper.R</code></a></small>
      <div class="d-none name"><code>IntegrateSpatialEcoTyper.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Integrate Multiple Spatial Transcriptomics Datasets to Identify Conserved Spatial Ecotypes</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">IntegrateSpatialEcoTyper</span><span class="op">(</span></span>
<span>  <span class="va">SpatialEcoTyper_list</span>,</span>
<span>  <span class="va">data_list</span>,</span>
<span>  outdir <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  normalization.method <span class="op">=</span> <span class="st">"None"</span>,</span>
<span>  nmf_ranks <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  nrun.per.rank <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  min.coph <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  nfeatures <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  min.features <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  Region <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subresolution <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  minibatch <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spatialecotyper-list">SpatialEcoTyper_list<a class="anchor" aria-label="anchor" href="#arg-spatialecotyper-list"></a></dt>
<dd><p>A named list of SpatialEcoTyper results, each item
represents a list returned from the <code>SpatialEcoTyper</code> function.</p></dd>


<dt id="arg-data-list">data_list<a class="anchor" aria-label="anchor" href="#arg-data-list"></a></dt>
<dd><p>A named list of expression matrices where each matrix represents
gene expression data used for the SpatialEcoTyper analysis. The list name should match
that of <code>SpatialEcoTyper_list</code>.</p></dd>


<dt id="arg-outdir">outdir<a class="anchor" aria-label="anchor" href="#arg-outdir"></a></dt>
<dd><p>Directory where the results will be saved. Defaults to the current
directory with a subdirectory named "SpatialEcoTyper_results_" followed by the current date.</p></dd>


<dt id="arg-normalization-method">normalization.method<a class="anchor" aria-label="anchor" href="#arg-normalization-method"></a></dt>
<dd><p>Method for normalizing the expression data. Options
include "None" (default), "SCT", or other methods compatible with Seurat's `NormalizeData` function.</p></dd>


<dt id="arg-nmf-ranks">nmf_ranks<a class="anchor" aria-label="anchor" href="#arg-nmf-ranks"></a></dt>
<dd><p>Integer or a vector specifying the number of clusters (10 by default).
When an integer vector is supplied, the function will test all supplied numbers and
select the optimal number, which takes time.</p></dd>


<dt id="arg-nrun-per-rank">nrun.per.rank<a class="anchor" aria-label="anchor" href="#arg-nrun-per-rank"></a></dt>
<dd><p>An integer specifying the the number of runs per rank for NMF (default: 30).</p></dd>


<dt id="arg-min-coph">min.coph<a class="anchor" aria-label="anchor" href="#arg-min-coph"></a></dt>
<dd><p>Numeric specifying the minimum cophenetic coefficient required for a rank to be optimal.</p></dd>


<dt id="arg-nfeatures">nfeatures<a class="anchor" aria-label="anchor" href="#arg-nfeatures"></a></dt>
<dd><p>An integer specifying the maximum number of top variable genes to select for each cell type.</p></dd>


<dt id="arg-min-features">min.features<a class="anchor" aria-label="anchor" href="#arg-min-features"></a></dt>
<dd><p>An integer specifying the minimum number of shared features (genes) required across samples.</p></dd>


<dt id="arg-region">Region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>A character string specifying the column name in metadata data frames
containing region annotations (default: NULL). Pathologist annotation is recommended if available.</p></dd>


<dt id="arg-subresolution">subresolution<a class="anchor" aria-label="anchor" href="#arg-subresolution"></a></dt>
<dd><p>A numeric specifying the resolution for clustering within each sample.</p></dd>


<dt id="arg-minibatch">minibatch<a class="anchor" aria-label="anchor" href="#arg-minibatch"></a></dt>
<dd><p>Integer specifying the number of columns to process in each minibatch in the SNF analysis.
Default is 5000. This option splits the matrix into smaller chunks (minibatch), thus reducing memory usage.</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>An integer specifying the number of cores for parallel processing. Default is 1.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>An integer used to seed the random number generator for NMF analysis.</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># See https://digitalcytometry.github.io/spatialecotyper/docs/articles/Integration.html</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

