<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Train SE Deconvolution Model — NMFGenerateW • SpatialEcoTyper</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Train SE Deconvolution Model — NMFGenerateW"><meta name="description" content="This function trains a Non-negative Matrix Factorization (NMF) model for SE deconvolution
based on given spatial ecotype fractions and gene expression matrix. Prior to NMF, each gene
is scaled to mean 0 and unit variance. To satisfy the non-negativity requirement of NMF,
the expression matrix is processed using posneg transformation, which converts the expression
matrix into two matrices, one containing only positive values and the other containing only
negative values with the sign inverted. The two matrices are subsequently concatenated to
produce the training data."><meta property="og:description" content="This function trains a Non-negative Matrix Factorization (NMF) model for SE deconvolution
based on given spatial ecotype fractions and gene expression matrix. Prior to NMF, each gene
is scaled to mean 0 and unit variance. To satisfy the non-negativity requirement of NMF,
the expression matrix is processed using posneg transformation, which converts the expression
matrix into two matrices, one containing only positive values and the other containing only
negative values with the sign inverted. The two matrices are subsequently concatenated to
produce the training data."><meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Train SE Deconvolution Model</h1>
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/R/NMFGenerateW.R" class="external-link"><code>R/NMFGenerateW.R</code></a></small>
      <div class="d-none name"><code>NMFGenerateW.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function trains a Non-negative Matrix Factorization (NMF) model for SE deconvolution
based on given spatial ecotype fractions and gene expression matrix. Prior to NMF, each gene
is scaled to mean 0 and unit variance. To satisfy the non-negativity requirement of NMF,
the expression matrix is processed using posneg transformation, which converts the expression
matrix into two matrices, one containing only positive values and the other containing only
negative values with the sign inverted. The two matrices are subsequently concatenated to
produce the training data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">NMFGenerateW</span><span class="op">(</span></span>
<span>  <span class="va">Fracs</span>,</span>
<span>  <span class="va">ExpMat</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nfeature <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  nfeature.per.se <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"brunet"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fracs">Fracs<a class="anchor" aria-label="anchor" href="#arg-fracs"></a></dt>
<dd><p>A fraction matrix, with rows as samples.</p></dd>


<dt id="arg-expmat">ExpMat<a class="anchor" aria-label="anchor" href="#arg-expmat"></a></dt>
<dd><p>A gene expression matrix with genes in rows and samples in columns.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Logical indicating whether to scale the gene expression matrix. Default is TRUE.</p></dd>


<dt id="arg-nfeature">nfeature<a class="anchor" aria-label="anchor" href="#arg-nfeature"></a></dt>
<dd><p>Integer specifying the top variable features for training the models (default: 2000).</p></dd>


<dt id="arg-nfeature-per-se">nfeature.per.se<a class="anchor" aria-label="anchor" href="#arg-nfeature-per-se"></a></dt>
<dd><p>Integer specifying the maximal number of features to select for each SE (default: 50).</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A character string specifying the NMF method to use. Default is "brunet".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix containing the basis NMF W matrix with rows as features and columns as SEs.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

