<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Recovery of SEs Using Pretrained NMF Models — RecoverSE • SpatialEcoTyper</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recovery of SEs Using Pretrained NMF Models — RecoverSE"><meta name="description" content="This function can recover SEs from scRNA-seq data or single-cell spatial data by assigning each single cell to an SE or NonSE."><meta property="og:description" content="This function can recover SEs from scRNA-seq data or single-cell spatial data by assigning each single cell to an SE or NonSE."><meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Recovery of SEs Using Pretrained NMF Models</h1>
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/R/RecoverSE.R" class="external-link"><code>R/RecoverSE.R</code></a></small>
      <div class="d-none name"><code>RecoverSE.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function can recover SEs from scRNA-seq data or single-cell spatial data by assigning each single cell to an SE or NonSE.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RecoverSE</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  celltypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ncell.per.run <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  Ws <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  se_results <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dat">dat<a class="anchor" aria-label="anchor" href="#arg-dat"></a></dt>
<dd><p>A numeric (sparse) matrix of gene expression data, from single-cell spatial
transcriptomics or scRNA-seq data.</p></dd>


<dt id="arg-celltypes">celltypes<a class="anchor" aria-label="anchor" href="#arg-celltypes"></a></dt>
<dd><p>Character vector specifying the cell type annotations for cells included in the gene expression `dat`.
If you're using the default model, cell types like B, CD4T, CD8T, NK, Plasma, Macrophage, DC, Fibroblast, Endothelial are expected.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Logical indicating whether to perform unit-variance normalization (default: TRUE). Change it with caution.</p></dd>


<dt id="arg-ncell-per-run">ncell.per.run<a class="anchor" aria-label="anchor" href="#arg-ncell-per-run"></a></dt>
<dd><p>Integer specifying the maximum number of cells per NMF prediction run to avoid memory issues.</p></dd>


<dt id="arg-ws">Ws<a class="anchor" aria-label="anchor" href="#arg-ws"></a></dt>
<dd><p>A list of cell-type-specific W matrices used to recover SE-specific
cell states. Each element in the list should be named after the corresponding
cell type and contain a W matrix from an NMF model.</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>Integer specifying the number of CPU cores to use for parallel processing (default: 1).</p></dd>


<dt id="arg-se-results">se_results<a class="anchor" aria-label="anchor" href="#arg-se-results"></a></dt>
<dd><p>A list including a seurat object and a metadata with spatial
cluster annotations (SE column) returned by SpatialEcoTyper. When supplied, the `dat`
should be single cell gene expression data used for the SpatialEcoTyper analysis.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with four columns, including CID (cell id), CellType (cell type),
SE (spatial ecotype), and PredScore (prediction probability for the assigned SE).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># see https://digitalcytometry.github.io/spatialecotyper/articles/Recovery_scRNA.html</span></span></span>
<span class="r-in"><span><span class="co"># see https://digitalcytometry.github.io/spatialecotyper/articles/Recovery_Spatial.html</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

