<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Identify Spatial EcoTypes from Single-cell Spatial Data (A Single Sample) — SpatialEcoTyper • SpatialEcoTyper</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify Spatial EcoTypes from Single-cell Spatial Data (A Single Sample) — SpatialEcoTyper"><meta name="description" content="This function identifies spatially distinct cellular ecosystems (SE) from a single sample."><meta property="og:description" content="This function identifies spatially distinct cellular ecosystems (SE) from a single sample."><meta property="og:image" content="https://digitalcytometry.github.io/spatialecotyper/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpatialEcoTyper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.0.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/SingleSample.html">Discovery of Spatial Ecotypes from a Single Sample</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Discovery of Spatial Ecotypes from Multiple Samples</a></li>
    <li><a class="dropdown-item" href="../articles/TrainRecoveryModels.html">Development of NMF Models for Spatial Ecotype Recovery</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_scRNA.html">Recovery of Spatial Ecotypes from Single-Cell Gene Expression Data</a></li>
    <li><a class="dropdown-item" href="../articles/Recovery_Bulk.html">Recovery of Spatial Ecotypes from Bulk Gene Expression Data</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">R functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/digitalcytometry/spatialecotyper" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identify Spatial EcoTypes from Single-cell Spatial Data (A Single Sample)</h1>
      <small class="dont-index">Source: <a href="https://github.com/digitalcytometry/spatialecotyper/blob/main/R/SpatialEcoTyper.R" class="external-link"><code>R/SpatialEcoTyper.R</code></a></small>
      <div class="d-none name"><code>SpatialEcoTyper.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function identifies spatially distinct cellular ecosystems (SE) from a single sample.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SpatialEcoTyper</span><span class="op">(</span></span>
<span>  <span class="va">normdata</span>,</span>
<span>  <span class="va">metadata</span>,</span>
<span>  outprefix <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>  radius <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  nfeatures <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  min.cts.per.region <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  npcs <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min.features <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  iterations <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minibatch <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  grid.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">radius</span> <span class="op">*</span> <span class="fl">1.4</span><span class="op">)</span>,</span>
<span>  filter.region.by.celltypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  k.sn <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  dropcell <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-normdata">normdata<a class="anchor" aria-label="anchor" href="#arg-normdata"></a></dt>
<dd><p>A matrix representing normalized gene expression data,
where rows correspond to genes and columns correspond to cells.</p></dd>


<dt id="arg-metadata">metadata<a class="anchor" aria-label="anchor" href="#arg-metadata"></a></dt>
<dd><p>A data frame containing metadata associated with each cell.
Must include spatial coordinates (e.g., X and Y) as well as cell type annotations.
The row names of the <code>metadata</code> must match the column names of the <code>normdata</code>.</p></dd>


<dt id="arg-outprefix">outprefix<a class="anchor" aria-label="anchor" href="#arg-outprefix"></a></dt>
<dd><p>Character string specifying the prefix for output file names.</p></dd>


<dt id="arg-radius">radius<a class="anchor" aria-label="anchor" href="#arg-radius"></a></dt>
<dd><p>Numeric specifying the radius (in the same units as spatial coordinates)
for defining spatial neighborhoods around each cell. Default is 50.</p></dd>


<dt id="arg-resolution">resolution<a class="anchor" aria-label="anchor" href="#arg-resolution"></a></dt>
<dd><p>Numeric specifying the resolution for Louvain clustering (default: 0.5).</p></dd>


<dt id="arg-nfeatures">nfeatures<a class="anchor" aria-label="anchor" href="#arg-nfeatures"></a></dt>
<dd><p>Integer specifying the number of top variable features (default: 500) used for the analysis.</p></dd>


<dt id="arg-min-cts-per-region">min.cts.per.region<a class="anchor" aria-label="anchor" href="#arg-min-cts-per-region"></a></dt>
<dd><p>Integer specifying the minimum number of cell types required for a spatial neighborhood.</p></dd>


<dt id="arg-npcs">npcs<a class="anchor" aria-label="anchor" href="#arg-npcs"></a></dt>
<dd><p>Integer specifying the number of principal components (PCs) (default: 20).</p></dd>


<dt id="arg-min-cells">min.cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>Minimum number of cells / spatial-meta-cells (default: 5)  expressing a feature/gene.</p></dd>


<dt id="arg-min-features">min.features<a class="anchor" aria-label="anchor" href="#arg-min-features"></a></dt>
<dd><p>Minimum number of features (default: 10) detected in a cell / spatial-meta-cell.</p></dd>


<dt id="arg-iterations">iterations<a class="anchor" aria-label="anchor" href="#arg-iterations"></a></dt>
<dd><p>Integer specifying the number of iterations (default: 10) for SNF analysis.</p></dd>


<dt id="arg-minibatch">minibatch<a class="anchor" aria-label="anchor" href="#arg-minibatch"></a></dt>
<dd><p>Integer specifying the number of columns to process in each minibatch in the SNF analysis.
Default is 5000. This option splits the matrix into smaller chunks (minibatch), thus reducing memory usage.</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>Integer specifying the number of CPU cores to use for parallel processing.</p></dd>


<dt id="arg-grid-size">grid.size<a class="anchor" aria-label="anchor" href="#arg-grid-size"></a></dt>
<dd><p>Numeric specifying the grid size for spatial discretization of coordinates. By default,
this size is determined based on the specified radius round(radius*1.4 µm). Increasing the grid.size will
downsample spatial neighborhoods and expedite the analysis, while it might eliminate cells located between bins
from the SE discovery analysis.</p></dd>


<dt id="arg-filter-region-by-celltypes">filter.region.by.celltypes<a class="anchor" aria-label="anchor" href="#arg-filter-region-by-celltypes"></a></dt>
<dd><p>A character vector specifying the cell types to include in the analysis.
Only spatial neighborhoods that contain at least one of the specified cell types will be analyzed, while regions
lacking these cell types will be excluded from the SE discovery process. If NULL, all spatial neighborhoods will
be included, regardless of cell type composition.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Integer specifying the number of spatial nearest neighbors (default: 20) used to construct spatial meta-cells.</p></dd>


<dt id="arg-k-sn">k.sn<a class="anchor" aria-label="anchor" href="#arg-k-sn"></a></dt>
<dd><p>Integer specifying the number of nearest neighbors (default: 50) for constructing similarity network.</p></dd>


<dt id="arg-dropcell">dropcell<a class="anchor" aria-label="anchor" href="#arg-dropcell"></a></dt>
<dd><p>Logical. If TRUE, cells that cannot be assigned to any spatial
ecotype (outside the radius) will be removed from the returned metadata. Default is TRUE.</p>
<p>#' @return A list containing two elements:</p><dl><dt>obj</dt>
<dd><p>A seurat object constructed from fused similarity network of sptial neighborhoods</p></dd>

  <dt>metadata</dt>
<dd><p>Updated <code>metadata</code>, with a new column (`SE`) added</p></dd>


</dl></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># See https://digitalcytometry.github.io/spatialecotyper/docs/articles/SingleSample.html</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://henrikbengtsson.github.io/R.utils/" class="external-link">R.utils</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://digitalcytometry.github/spatialecotyper" class="external-link">SpatialEcoTyper</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_deauth.html" class="external-link">drive_deauth</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># Disable Google sign-in requirement</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"13Rc5Rsu8jbnEYYfUse-xQ7ges51LcI7n"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in drive_download(as_id("13Rc5Rsu8jbnEYYfUse-xQ7ges51LcI7n"), "HumanMelanomaPatient1_subset_counts.tsv.gz"):</span> Local `path` already exists and overwrite is `FALSE`:</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> <span style="color: #0000BB;">HumanMelanomaPatient1_subset_counts.tsv.gz</span></span>
<span class="r-in"><span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_download.html" class="external-link">drive_download</a></span><span class="op">(</span><span class="fu"><a href="https://googledrive.tidyverse.org/reference/drive_id.html" class="external-link">as_id</a></span><span class="op">(</span><span class="st">"12xcZNhpT-xbhcG8kX1QAdTeM9TKeFAUW"</span><span class="op">)</span>, <span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in drive_download(as_id("12xcZNhpT-xbhcG8kX1QAdTeM9TKeFAUW"), "HumanMelanomaPatient1_subset_scmeta.tsv"):</span> Local `path` already exists and overwrite is `FALSE`:</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">•</span> <span style="color: #0000BB;">HumanMelanomaPatient1_subset_scmeta.tsv</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load single-cell gene expression matrix. Rows are genes, columns are cells.</span></span></span>
<span class="r-in"><span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_counts.tsv.gz"</span>,</span></span>
<span class="r-in"><span>                sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, data.table <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co"># set genes as row names</span></span></span>
<span class="r-in"><span><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">normdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">normdata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          HumanMelanomaPatient1__cell_3655 HumanMelanomaPatient1__cell_3657</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PDK4                             0.000000                         4.221184</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TNFRSF17                         0.000000                         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ICAM3                            0.000000                         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FAP                              3.552438                         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GZMB                             0.000000                         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TSC2                             0.000000                         0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          HumanMelanomaPatient1__cell_3658 HumanMelanomaPatient1__cell_3660</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PDK4                             3.208945                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TNFRSF17                         0.000000                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ICAM3                            0.000000                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FAP                              0.000000                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GZMB                             0.000000                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TSC2                             0.000000                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          HumanMelanomaPatient1__cell_3661</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PDK4                                    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TNFRSF17                                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ICAM3                                   0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FAP                                     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GZMB                                    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TSC2                                    0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Load single-cell metadata. Three columns are required, including X, Y, and CellType. Others are optional.</span></span></span>
<span class="r-in"><span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"HumanMelanomaPatient1_subset_scmeta.tsv"</span>,</span></span>
<span class="r-in"><span>                     sep <span class="op">=</span> <span class="st">"\t"</span>,header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">scmeta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>, <span class="op">]</span> <span class="co"># match the cell ids in scdata and scmeta</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                         X         Y   CellType CellTypeName</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3657 1942.480 -6369.602 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3658 1963.007 -6374.026 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3660 1981.600 -6372.266 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  Region Dist2Interface</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3655 Stroma      -883.1752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3657 Stroma      -894.8463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3658 Stroma      -904.1115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3660 Stroma      -907.8909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3661 Stroma      -874.2712</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3663 Stroma      -903.6559</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">se_results</span> <span class="op">&lt;-</span> <span class="fu">SpatialEcoTyper</span><span class="op">(</span><span class="va">normdata</span>, <span class="va">scmeta</span>,</span></span>
<span class="r-in"><span>                              outprefix <span class="op">=</span> <span class="st">"Melanoma1_subset"</span>,</span></span>
<span class="r-in"><span>                              radius <span class="op">=</span> <span class="fl">50</span>, ncores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:17.608242 Remove 88 genes expressed in fewer than 5 cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:18.286728 Construct spatial meta cells for each cell type</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 		Total spatial neighborhoods: 2315</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 		Total spatial meta cells: 7711</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:26.833969 PCA for each cell type</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:28.097188 Construct cell-type-specific similarity network</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 		Remove 446 spatial neighborhoods with fewer than 2 cell types </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:28.664179 Similarity network fusion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:28.664702 Normalize networks ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:28.753693 Calculate the local transition matrix ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:43:36.205797 Perform the diffusion ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:43:36.206637 Iteration: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:43:40.068319 Iteration: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:43:48.739085 Iteration: 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:00.653146 Iteration: 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:08.782312 Iteration: 5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:18.374433 Iteration: 6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:28.224813 Iteration: 7</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:38.996738 Iteration: 8</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:51.346514 Iteration: 9</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	2026-01-21 09:44:59.655105 Iteration: 10</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:45:12.959869 Create Seurat object and perform clustering analysis</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Data is of class dgeMatrix. Coercing to dgCMatrix.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> This message will be shown once per session</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2026-01-21 09:45:19.614424 The Seurat object is saved into Melanoma1_subset_SpatialEcoTyper_results.rds</span>
<span class="r-in"><span><span class="co"># Extract the Seurat object and updated single-cell metadata</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">$</span><span class="va">obj</span> <span class="co"># A Seurat object</span></span></span>
<span class="r-in"><span><span class="va">obj</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class Seurat </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1869 features across 1869 samples within 1 assay </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Active assay: RNA (1869 features, 1869 variable features)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3 layers present: counts, data, scale.data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2 dimensional reductions calculated: pca, umap</span>
<span class="r-in"><span><span class="va">scmeta</span> <span class="op">&lt;-</span> <span class="va">se_results</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">SE</span><span class="op">)</span> <span class="co"># Updated single-cell meta data, with SE annotation added</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scmeta</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                         X         Y   CellType CellTypeName</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3655 1894.706 -6367.766 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3661 1742.939 -6374.851 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3663 1921.683 -6383.309 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3664 1706.253 -6383.428 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3666 1761.403 -6387.970 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3668 1841.514 -6389.663 Fibroblast  Fibroblasts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  Region Dist2Interface  SE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3655 Stroma      -883.1752 SE0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3661 Stroma      -874.2712 SE0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3663 Stroma      -903.6559 SE0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3664 Stroma      -880.4758 SE0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3666 Stroma      -888.6271 SE0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HumanMelanomaPatient1__cell_3668 Stroma      -896.8484 SE0</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://profiles.stanford.edu/wubing-zhang" class="external-link">Wubing Zhang</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

